<!DOCTYPE html>


<html theme="dark" showBanner="true" hasBanner="false" > 
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.5.1/css/all.min.css" integrity="sha512-DTOQO9RWCH3ppGqcWaEA1BIZOC6xxalwEsw9c2QQeAIftl+Vegovlnee1c9QX4TctnWMn13TZye+giMm8e2LwA==" crossorigin="anonymous" referrerpolicy="no-referrer" />
<script src="/blog/js/color.global.min.js" ></script>
<script src="/blog/js/load-settings.js" ></script>
<head>
  <meta charset="utf-8">
  
  
  

  
  <title>单细胞数据挖掘 | kusurin.blog</title>
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">

  <meta name="description" content="案例里用了《Glioblastoma cell differentiation trajectory predicts the immunotherapy response and overall survival of patients》，主要是使用integrated scRNA-seq（综合（？）单细胞RNA测序）和bulk RNA-seq（整体（？）RNA测序）来找出GBM分化相关">
<meta property="og:type" content="article">
<meta property="og:title" content="单细胞数据挖掘">
<meta property="og:url" content="https://kusurin.github.io/blog/2024/05/22/single-cell-data-mining/index.html">
<meta property="og:site_name" content="kusurin.blog">
<meta property="og:description" content="案例里用了《Glioblastoma cell differentiation trajectory predicts the immunotherapy response and overall survival of patients》，主要是使用integrated scRNA-seq（综合（？）单细胞RNA测序）和bulk RNA-seq（整体（？）RNA测序）来找出GBM分化相关">
<meta property="og:locale" content="zh_CN">
<meta property="og:image" content="https://kusurin.github.io/blog/images/square_hole.png">
<meta property="og:image" content="https://cdn-ak.f.st-hatena.com/images/fotolife/k/kimoppy126/20180902/20180902233845.png">
<meta property="article:published_time" content="2024-05-22T15:14:26.000Z">
<meta property="article:modified_time" content="2024-08-12T12:23:29.875Z">
<meta property="article:author" content="kusurin">
<meta property="article:tag" content="生信">
<meta property="article:tag" content="R based">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://kusurin.github.io/blog/images/square_hole.png">
  
    <link rel="alternate" href="/blog/atom.xml" title="kusurin.blog" type="application/atom+xml">
  
  
    <link rel="icon" media="(prefers-color-scheme: light)" href="/blog/images/favicon-light-32.png" sizes="32x32">
    <link rel="icon" media="(prefers-color-scheme: light)" href="/blog/images/favicon-light-128.png" sizes="128x128">
    <link rel="icon" media="(prefers-color-scheme: light)" href="/blog/images/favicon-light-180.png" sizes="180x180">
    <link rel="icon" media="(prefers-color-scheme: light)" href="/blog/images/favicon-light-192.png" sizes="192x192">
    <link rel="icon" media="(prefers-color-scheme: dark)" href="/blog/images/favicon-dark-32.png" sizes="32x32">
    <link rel="icon" media="(prefers-color-scheme: dark)" href="/blog/images/favicon-dark-128.png" sizes="128x128">
    <link rel="icon" media="(prefers-color-scheme: dark)" href="/blog/images/favicon-dark-180.png" sizes="180x180">
    <link rel="icon" media="(prefers-color-scheme: dark)" href="/blog/images/favicon-dark-192.png" sizes="192x192">
  
  
<link rel="stylesheet" href="/blog/css/style.css">

<meta name="generator" content="Hexo 6.3.0"><style>mjx-container[jax="SVG"] {
  direction: ltr;
}

mjx-container[jax="SVG"] > svg {
  overflow: visible;
}

mjx-container[jax="SVG"][display="true"] {
  display: block;
  text-align: center;
  margin: 1em 0;
}

mjx-container[jax="SVG"][justify="left"] {
  text-align: left;
}

mjx-container[jax="SVG"][justify="right"] {
  text-align: right;
}

g[data-mml-node="merror"] > g {
  fill: red;
  stroke: red;
}

g[data-mml-node="merror"] > rect[data-background] {
  fill: yellow;
  stroke: none;
}

g[data-mml-node="mtable"] > line[data-line] {
  stroke-width: 70px;
  fill: none;
}

g[data-mml-node="mtable"] > rect[data-frame] {
  stroke-width: 70px;
  fill: none;
}

g[data-mml-node="mtable"] > .mjx-dashed {
  stroke-dasharray: 140;
}

g[data-mml-node="mtable"] > .mjx-dotted {
  stroke-linecap: round;
  stroke-dasharray: 0,140;
}

g[data-mml-node="mtable"] > svg {
  overflow: visible;
}

[jax="SVG"] mjx-tool {
  display: inline-block;
  position: relative;
  width: 0;
  height: 0;
}

[jax="SVG"] mjx-tool > mjx-tip {
  position: absolute;
  top: 0;
  left: 0;
}

mjx-tool > mjx-tip {
  display: inline-block;
  padding: .2em;
  border: 1px solid #888;
  font-size: 70%;
  background-color: #F8F8F8;
  color: black;
  box-shadow: 2px 2px 5px #AAAAAA;
}

g[data-mml-node="maction"][data-toggle] {
  cursor: pointer;
}

mjx-status {
  display: block;
  position: fixed;
  left: 1em;
  bottom: 1em;
  min-width: 25%;
  padding: .2em .4em;
  border: 1px solid #888;
  font-size: 90%;
  background-color: #F8F8F8;
  color: black;
}

foreignObject[data-mjx-xml] {
  font-family: initial;
  line-height: normal;
  overflow: visible;
}

.MathJax path {
  stroke-width: 3;
}

mjx-container[display="true"] {
  overflow: auto hidden;
}

mjx-container[display="true"] + br {
  display: none;
}
</style></head>

<body>
  
   
  <div id="main-grid" class="shadow   ">
    <div id="nav" class=""  >
      <navbar id="navbar">
  <nav id="title-nav">
    <a href="/blog/">
      <div id="vivia-logo">
        <div class="dot"></div>
        <div class="dot"></div>
        <div class="dot"></div>
        <div class="dot"></div>
      </div>
      <div>kusurin.blog </div>
    </a>
  </nav>
  <nav id="main-nav">
    
      <a class="main-nav-link" href="/blog/">Home</a>
    
      <a class="main-nav-link" href="/blog/archives">Archives</a>
    
      <a class="main-nav-link" href="/blog/about">About</a>
    
  </nav>
  <nav id="sub-nav">
    <a id="theme-btn" class="nav-icon">
      <span class="light-mode-icon"><svg xmlns="http://www.w3.org/2000/svg" height="20" viewBox="0 -960 960 960" width="20"><path d="M438.5-829.913v-48q0-17.452 11.963-29.476 11.964-12.024 29.326-12.024 17.363 0 29.537 12.024t12.174 29.476v48q0 17.452-11.963 29.476-11.964 12.024-29.326 12.024-17.363 0-29.537-12.024T438.5-829.913Zm0 747.826v-48q0-17.452 11.963-29.476 11.964-12.024 29.326-12.024 17.363 0 29.537 12.024t12.174 29.476v48q0 17.452-11.963 29.476-11.964 12.024-29.326 12.024-17.363 0-29.537-12.024T438.5-82.087ZM877.913-438.5h-48q-17.452 0-29.476-11.963-12.024-11.964-12.024-29.326 0-17.363 12.024-29.537t29.476-12.174h48q17.452 0 29.476 11.963 12.024 11.964 12.024 29.326 0 17.363-12.024 29.537T877.913-438.5Zm-747.826 0h-48q-17.452 0-29.476-11.963-12.024-11.964-12.024-29.326 0-17.363 12.024-29.537T82.087-521.5h48q17.452 0 29.476 11.963 12.024 11.964 12.024 29.326 0 17.363-12.024 29.537T130.087-438.5Zm660.174-290.87-34.239 32q-12.913 12.674-29.565 12.174-16.653-.5-29.327-13.174-12.674-12.673-12.554-28.826.12-16.152 12.794-28.826l33-35q12.913-12.674 30.454-12.674t30.163 12.847q12.709 12.846 12.328 30.826-.38 17.98-13.054 30.653ZM262.63-203.978l-32 34q-12.913 12.674-30.454 12.674t-30.163-12.847q-12.709-12.846-12.328-30.826.38-17.98 13.054-30.653l33.239-31q12.913-12.674 29.565-12.174 16.653.5 29.327 13.174 12.674 12.673 12.554 28.826-.12 16.152-12.794 28.826Zm466.74 33.239-32-33.239q-12.674-12.913-12.174-29.565.5-16.653 13.174-29.327 12.673-12.674 28.826-13.054 16.152-.38 28.826 12.294l35 33q12.674 12.913 12.674 30.454t-12.847 30.163q-12.846 12.709-30.826 12.328-17.98-.38-30.653-13.054ZM203.978-697.37l-34-33q-12.674-12.913-13.174-29.945-.5-17.033 12.174-29.707t31.326-13.293q18.653-.62 31.326 13.054l32 34.239q11.674 12.913 11.174 29.565-.5 16.653-13.174 29.327-12.673 12.674-28.826 12.554-16.152-.12-28.826-12.794ZM480-240q-100 0-170-70t-70-170q0-100 70-170t170-70q100 0 170 70t70 170q0 100-70 170t-170 70Zm-.247-82q65.703 0 111.475-46.272Q637-414.544 637-480.247t-45.525-111.228Q545.95-637 480.247-637t-111.475 45.525Q323-545.95 323-480.247t45.525 111.975Q414.05-322 479.753-322ZM481-481Z"/></svg></span>
      <span class="dark-mode-icon"><svg xmlns="http://www.w3.org/2000/svg" height="20" viewBox="0 -960 960 960" width="20"><path d="M480.239-116.413q-152.63 0-258.228-105.478Q116.413-327.37 116.413-480q0-130.935 77.739-227.435t206.304-125.043q43.022-9.631 63.87 10.869t3.478 62.805q-8.891 22.043-14.315 44.463-5.424 22.42-5.424 46.689 0 91.694 64.326 155.879 64.325 64.186 156.218 64.186 24.369 0 46.978-4.946 22.609-4.945 44.413-14.076 42.826-17.369 62.967 1.142 20.142 18.511 10.511 61.054Q807.174-280 712.63-198.206q-94.543 81.793-232.391 81.793Zm0-95q79.783 0 143.337-40.217 63.554-40.218 95.793-108.283-15.608 4.044-31.097 5.326-15.49 1.283-31.859.805-123.706-4.066-210.777-90.539-87.071-86.473-91.614-212.092-.24-16.369.923-31.978 1.164-15.609 5.446-30.978-67.826 32.478-108.282 96.152Q211.652-559.543 211.652-480q0 111.929 78.329 190.258 78.329 78.329 190.258 78.329ZM466.13-465.891Z"/></svg></span>
    </a>
    
      <a id="nav-rss-link" class="nav-icon mobile-hide" href="/blog/atom.xml" title="RSS 订阅">
        <svg xmlns="http://www.w3.org/2000/svg" height="20" viewBox="0 -960 960 960" width="20"><path d="M198-120q-25.846 0-44.23-18.384-18.384-18.385-18.384-44.23 0-25.846 18.384-44.23 18.384-18.385 44.23-18.385 25.846 0 44.23 18.385 18.384 18.384 18.384 44.23 0 25.845-18.384 44.23Q223.846-120 198-120Zm538.385 0q-18.846 0-32.923-13.769-14.076-13.769-15.922-33.23-8.692-100.616-51.077-188.654-42.385-88.039-109.885-155.539-67.5-67.501-155.539-109.885Q283-663.462 182.385-672.154q-19.461-1.846-33.23-16.23-13.769-14.385-13.769-33.846t14.076-32.922q14.077-13.461 32.923-12.23 120.076 8.692 226.038 58.768 105.961 50.077 185.73 129.846 79.769 79.769 129.846 185.731 50.077 105.961 58.769 226.038 1.231 18.846-12.538 32.922Q756.461-120 736.385-120Zm-252 0q-18.231 0-32.423-13.461t-18.653-33.538Q418.155-264.23 348.886-333.5q-69.27-69.27-166.501-84.423-20.077-4.462-33.538-18.961-13.461-14.5-13.461-33.346 0-19.076 13.884-33.23 13.884-14.153 33.115-10.922 136.769 15.384 234.384 112.999 97.615 97.615 112.999 234.384 3.231 19.23-10.538 33.115Q505.461-120 484.385-120Z"/></svg>
      </a>
    
    <div id="nav-menu-btn" class="nav-icon">
      <svg xmlns="http://www.w3.org/2000/svg" height="20" viewBox="0 -960 960 960" width="20"><path d="M177.37-252.282q-17.453 0-29.477-11.964-12.024-11.963-12.024-29.326t12.024-29.537q12.024-12.174 29.477-12.174h605.26q17.453 0 29.477 11.964 12.024 11.963 12.024 29.326t-12.024 29.537q-12.024 12.174-29.477 12.174H177.37Zm0-186.218q-17.453 0-29.477-11.963-12.024-11.964-12.024-29.326 0-17.363 12.024-29.537T177.37-521.5h605.26q17.453 0 29.477 11.963 12.024 11.964 12.024 29.326 0 17.363-12.024 29.537T782.63-438.5H177.37Zm0-186.217q-17.453 0-29.477-11.964-12.024-11.963-12.024-29.326t12.024-29.537q12.024-12.174 29.477-12.174h605.26q17.453 0 29.477 11.964 12.024 11.963 12.024 29.326t-12.024 29.537q-12.024 12.174-29.477 12.174H177.37Z"/></svg>
    </div>
  </nav>
</navbar>
<div id="nav-dropdown" class="hidden">
  <div id="dropdown-link-list">
    
      <a class="nav-dropdown-link" href="/blog/">Home</a>
    
      <a class="nav-dropdown-link" href="/blog/archives">Archives</a>
    
      <a class="nav-dropdown-link" href="/blog/about">About</a>
    
    
      <a class="nav-dropdown-link" href="/blog/atom.xml" title="RSS 订阅">RSS</a>
     
    </div>
</div>
<script>
  let dropdownBtn = document.getElementById("nav-menu-btn");
  let dropdownEle = document.getElementById("nav-dropdown");
  dropdownBtn.onclick = function() {
    dropdownEle.classList.toggle("hidden");
  }
</script>
    </div>
    <div id="sidebar-wrapper">
      <sidebar id="sidebar">
  
    <div class="widget-wrap">
  <div class="info-card">
    <div class="avatar">
      
        <image src=/blog/images/profile_pic_2.jpg></image>
      
      <div class="img-dim"></div>
    </div>
    <div class="info">
      <div class="username">kusurin </div>
      <div class="dot"></div>
      <div class="subtitle">文字の中に沈む </div>
      <div class="link-list">
        
          <a class="link-btn" target="_blank" rel="noopener" href="https://kusurin.icu" title="HomePage"><i class="fa-solid fa-house"></i></a>
        
          <a class="link-btn" target="_blank" rel="noopener" href="https://twitter.com/Kusurin3" title="Twitter"><i class="fa-brands fa-twitter"></i></a>
        
          <a class="link-btn" target="_blank" rel="noopener" href="https://github.com/kusurin" title="GitHub"><i class="fa-brands fa-github"></i></a>
        
          <a class="link-btn" target="_blank" rel="noopener" href="https://www.zhihu.com/people/kusurin" title="知乎"><i class="fa-brands fa-zhihu"></i></a>
        
          <a class="link-btn" target="_blank" rel="noopener" href="https://space.bilibili.com/5404263" title="bilibili"><i class="fa-brands fa-bilibili"></i></a>
         
      </div>  
    </div>
  </div>
</div>

  
  <div class="sticky">
    
      


  <div class="widget-wrap">
    <div class="widget">
      <h3 class="widget-title">分类</h3>
      <div class="category-box">
            <a class="category-link" href="/blog/categories/CSS-tricks/">
                CSS-tricks
                <div class="category-count">5</div>
            </a>
        
            <a class="category-link" href="/blog/categories/%E6%9D%82%E8%B0%88/">
                杂谈
                <div class="category-count">7</div>
            </a>
        <div class="children"><div class="category-box">
            <a class="category-link" href="/blog/categories/%E6%9D%82%E8%B0%88/%E7%BD%91%E5%AE%89/">
                网安
                <div class="category-count">3</div>
            </a>
        </div></div>
            <a class="category-link" href="/blog/categories/%E8%B8%A9%E5%9D%91%E6%97%A5%E8%AE%B0/">
                踩坑日记
                <div class="category-count">2</div>
            </a>
        
            <a class="category-link" href="/blog/categories/%E7%AE%97%E6%B3%95/">
                算法
                <div class="category-count">1</div>
            </a>
        
            <a class="category-link" href="/blog/categories/%E7%94%9F%E4%BF%A1/">
                生信
                <div class="category-count">2</div>
            </a>
        </div>
    </div>
  </div>


    
      
  <div class="widget-wrap">
    <div class="widget">
      <h3 class="widget-title">标签</h3>
      <ul class="widget-tag-list" itemprop="keywords"><li class="widget-tag-list-item"><a class="widget-tag-list-link" href="/blog/tags/CSS/" rel="tag">CSS</a></li><li class="widget-tag-list-item"><a class="widget-tag-list-link" href="/blog/tags/FFT/" rel="tag">FFT</a></li><li class="widget-tag-list-item"><a class="widget-tag-list-link" href="/blog/tags/R-based/" rel="tag">R based</a></li><li class="widget-tag-list-item"><a class="widget-tag-list-link" href="/blog/tags/WPA2/" rel="tag">WPA2</a></li><li class="widget-tag-list-item"><a class="widget-tag-list-link" href="/blog/tags/%E6%8F%8F%E8%BE%B9/" rel="tag">描边</a></li><li class="widget-tag-list-item"><a class="widget-tag-list-link" href="/blog/tags/%E6%9D%82%E8%B0%88/" rel="tag">杂谈</a></li><li class="widget-tag-list-item"><a class="widget-tag-list-link" href="/blog/tags/%E6%A8%A1%E6%8B%9F%E4%BD%8D%E7%BD%AE/" rel="tag">模拟位置</a></li><li class="widget-tag-list-item"><a class="widget-tag-list-link" href="/blog/tags/%E7%94%9F%E4%BF%A1/" rel="tag">生信</a></li><li class="widget-tag-list-item"><a class="widget-tag-list-link" href="/blog/tags/%E7%AE%97%E6%B3%95/" rel="tag">算法</a></li><li class="widget-tag-list-item"><a class="widget-tag-list-link" href="/blog/tags/%E7%BD%91%E5%AE%89/" rel="tag">网安</a></li><li class="widget-tag-list-item"><a class="widget-tag-list-link" href="/blog/tags/%E8%9D%B4%E8%9D%B6%E5%8F%98%E6%8D%A2/" rel="tag">蝴蝶变换</a></li><li class="widget-tag-list-item"><a class="widget-tag-list-link" href="/blog/tags/%E8%B8%A9%E5%9D%91/" rel="tag">踩坑</a></li><li class="widget-tag-list-item"><a class="widget-tag-list-link" href="/blog/tags/%E8%BE%B9%E6%A1%86/" rel="tag">边框</a></li></ul>
    </div>
  </div>


    
  </div>
</sidebar>
    </div>
    <div id="content-body">
       


<article id="post-single-cell-data-mining" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  
    
   
  <div class="article-inner">
    <div class="article-main">
      <header class="article-header">
        
<div class="main-title-bar">
  <div class="main-title-dot"></div>
  
    
      <h1 class="p-name article-title" itemprop="headline name">
        单细胞数据挖掘
      </h1>
    
  
</div>

        <div class='meta-info-bar'>
          <div class="meta-info">
  <time class="dt-published" datetime="2024-05-22T15:14:26.000Z" itemprop="datePublished">2024-05-22</time>
</div>
          <div class="need-seperator meta-info">
            <div class="meta-cate-flex">
  
  <a class="meta-cate-link" href="/blog/categories/%E7%94%9F%E4%BF%A1/">生信</a>
   
</div>
  
          </div>
          <div class="wordcount need-seperator meta-info">
            5.3k 词 
          </div>
        </div>
        
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/blog/tags/R-based/" rel="tag">R based</a></li><li class="article-tag-list-item"><a class="article-tag-list-link" href="/blog/tags/%E7%94%9F%E4%BF%A1/" rel="tag">生信</a></li></ul>

      </header>
      <div class="e-content article-entry" itemprop="articleBody">
        
          <p>案例里用了《Glioblastoma cell differentiation trajectory predicts the
immunotherapy response and overall survival of
patients》，主要是使用integrated
scRNA-seq（综合（？）单细胞RNA测序）和bulk
RNA-seq（整体（？）RNA测序）来找出GBM分化相关基因然后预测患者总生存期之类的。</p>
<p>换一个类似的做实验，也不懂各种数据有什么差别，就选《Cell
differentiation trajectory predicts patient potential immunotherapy
response and prognosis in gastric cancer》这篇文章，主要是使用integrated
scRNA-seq和bulk
RNA-seq来找出GC分化相关基因然后预测患者总生存期之类的。（干的事都差不多，真是原原又神神啊）<br>
这个用的数据集是GSE112302和GSE84437，和案例类似的部分用的是GSE112302。</p>
<h2 id="实验原理">实验原理</h2>
<p>主要就是干了质控、基因可视化、降维与聚类、周期判断、double
droplet、细胞类型注释、轨迹分析。</p>
<h3 id="数据集">数据集</h3>
<p>这堆数据集就是一轴是细胞或一坨样本，一轴是基因种类，每个点代表某个细胞/样本的某个基因表达量。</p>
<h3 id="质控">质控</h3>
<p>质控主要就是过滤：</p>
<ul>
<li>低质量的细胞(比如基因少、线粒体基因多)</li>
<li>双细胞（两个细胞被捕获到同一个液滴中）</li>
<li>高ERCC基因（External RNA Controls
Consortium），<em>“是一组合成的RNA分子，它们的浓度已知，并且在测序实验中添加到样本中，以便可以用来校正技术变异和进行数据标准化”</em>，在所有单元格中的浓度应该是一致的，所以如果ERCC基因百分比高，那这个单元格的RNA可能较少。</li>
</ul>
<p><code>Seurat</code>包可以傻瓜实现。</p>
<p>案例里还提到了double droplet和周期判断。<br>
double droplet就是上面说的双细胞。<br>
而周期判断就是说，一些细胞处于分裂期，它们会高表达某些基因，可能会影响聚类，所以要去掉。<br>
<code>scran</code>包可以傻瓜实现。</p>
<h3 id="挑选高变异基因">挑选高变异基因</h3>
<p>highly Variable
gene是在一组样本或细胞中，表达水平变化大的基因，它们可能是驱动细胞异质性的关键。
<code>Seurat</code>包同样可以傻瓜实现。</p>
<h3 id="降维与聚类">降维与聚类</h3>
<p>降维机器学习里也有，<em>“将数据从原始的高维空间转换到低维空间，同时尽可能保留数据的变异性”</em>，总的来说目的就是去掉冗杂信息。一个方法就是PCA（主成分分析），能使得数据的方差尽可能大（保留变异性），而均方误差尽可能小（和原数据的差距尽可能小），欸网上看证明去吧。<br>
PCA对数据的尺度（即每个特征的平均值和标准差）非常敏感，如果不同的feature有不同的尺度（0-1vs0-1000），那么PCA的结果将主要受到尺度大的特征的影响，所以PCA之前要标准化。</p>
<p>聚类就是把数据分成几类，使得同一个类内数据的相似性尽可能大，同时不同类数据的差异性也尽可能地大。欸还是无监督学习，感兴趣看看吧。</p>
<p>案例里PCA之后聚类然后又用t-SNE降维，有人说是因为先用PCA去除噪声，利于准确聚类，然后用t-SNE更好地可视化（<em>“在低维空间中保留高维空间中的邻域结构，可以有效地揭示数据中的群体结构和模式”</em>
）。<br>
另外PCA是线性的，而t-SNE是非线性的。</p>
<style>
    #square-hole img {
        display:none;
        position:absolute;
        left:0;
        top:100%;
    }
    #square-hole:hover img {
        display:block;
    }
</style>
<p><code>Seurat</code>包还是可以傻瓜实现。<a id="square-hole" onfocus="" ontouchstart="" style="position:relative;display:inline-block;word-break:all;"><del>that's
right the square hole</del>
<img src="/blog/images/square_hole.png" style=""></a></p>
<h3 id="细胞类型注释">细胞类型注释</h3>
<p>字面意思，就比如把一些细胞标注成B细胞之类的甚至更精细一些，幼稚B细胞之类的。这靠的是一些特异性高的marker基因。
可以用<code>SingleR</code>包。</p>
<h3 id="轨迹分析">轨迹分析</h3>
<p>就是“构建细胞的变化轨迹，重塑细胞随时间的变化过程”，目的是绘制祖细胞到分化细胞的进展情况。<br>
可以用<code>Monocle</code>包。</p>
<hr>
<p><span style="color:#808080;"><em><del>本来这点东西一天下班时间就能搞完的，现在下来两三天了，不光是因为只有业余时间，还真是怠惰啊。</del></em></span></p>
<hr>
<h2 id="r语言实现">R语言实现</h2>
<p>代码详见<a target="_blank" rel="noopener" href="https://github.com/kusurin/single-cell-data-mining">single-cell-data-mining</a>。</p>
<h3 id="提取数据">提取数据</h3>
<p>怎么两个数据集形式差那么多，你们能不能统一一下（骂骂咧咧）。</p>
<p>先看GSE84465学习一下。<br>
案例里的GSE84465集合到了一个<code>.csv</code>里，列名是某样本的plate_id（微孔板）和Well（孔）（我还没做过湿实验啊），行名就是基因。<br>
然后还有一个SraRunTable.txt，刚刚还奇怪，上面的文件都没有分组信息，这个文件里就有，就是每个样本的一些元信息，比如组织类型、患者编号、对应的样本编号等。这个就是SRA里的元数据，进SRA
Run Selector里就能找到。<br>
所以这个SraRunTable.txt就是用来给上面那些sc RNA-seq数据分组。</p>
<p>我们要做的就是从SRA生成GSE对应的列名，并使用SRA的信息筛选出肿瘤组织，借助对应的列名筛选出GSE中的肿瘤组织。<br>
还有SRA记录着分组信息，也要一起筛选出来。</p>
<p>所以说拿到GSE112302也是一样的，就是GSE112302没有给全部合在一起的数据，需要我们自己合起来（附加列，也就是样本），然后用SRA筛选出肿瘤组织。</p>
<h3 id="质控-1">质控</h3>
<h4 id="创建seurat对象简单筛选细胞">创建<code>Seurat</code>对象，简单筛选细胞</h4>
<p>要先创建<code>Seurat</code>对象。<code>CreateSeuratObject()</code>函数的文档如下：</p>
<blockquote>
<ul>
<li><code>counts</code>：这是一个矩阵类对象，包含未经标准化的数据，其中单元格作为列，特征作为行，或者是一个<code>Assay</code>派生对象。</li>
<li><code>assay</code>：初始测定的名称。</li>
<li><code>meta.data</code>：要添加到<code>Seurat</code>对象的额外单元格级元数据。应该是一个数据框，其中行是单元格名称，列是额外的元数据字段。元数据中的行名称需要与计数矩阵的列名称匹配。</li>
<li><code>project</code>：<code>Seurat</code>对象的项目名称。</li>
<li><code>...</code>：传递给其他方法的参数。</li>
<li><code>min.cells</code>：至少在这么多细胞中检测到的特征。也会对计数矩阵进行子集。要重新引入排除的特征，使用较低的截止值创建一个新对象。</li>
<li><code>min.features</code>：至少在这么多特征中检测到的细胞。</li>
</ul>
</blockquote>
<p>（对了R语言的函数是这样调用的：<code>function_name(arg1 = value1, arg2 = value2, ...)</code>）</p>
<p>其实只要用到<code>counts</code>、<code>meta.data</code>、<code>min.cells</code>、<code>min.features</code>就行了。
这里<code>counts</code>就是数据，<code>meta.data</code>就是分组信息，用<code>data.frame()</code>上面说了是这样的：<code>...</code>是列（额外的元数据字段，这里就是病人编号，要用来分组嘛），<code>row.names</code>是行（样本编号）,要和<code>counts</code>的列名对应。<br>
然后后两个前者用来筛选表达的细胞的数量低的基因，后者用来筛选基因少的细胞，顺便还能过滤空细胞。</p>
<h4 id="筛选线粒体基因和ercc基因过多细胞">筛选线粒体基因和ERCC基因过多细胞</h4>
<p>然后还要筛选掉线粒体基因过多的细胞： <figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">scRNA<span class="punctuation">[[</span><span class="string">"percent.mt"</span><span class="punctuation">]</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> PercentageFeatureSet<span class="punctuation">(</span>scRNA<span class="punctuation">,</span> pattern <span class="operator">=</span> <span class="string">"^MT-"</span><span class="punctuation">)</span> <span class="comment">#先全部算了比较好，PercentageFeatureSet()计算满足给定模式的特征在每个单元格中的百分比，后面那个参数就是正则（真是先进啊（大嘘））</span></span><br><span class="line">scRNA <span class="operator">&lt;-</span> subset<span class="punctuation">(</span>scRNA<span class="punctuation">,</span> subset <span class="operator">=</span> percent.mt <span class="operator">&lt;</span> <span class="number">5</span><span class="punctuation">)</span> <span class="comment">#案例取5%，subset()就是Array.filter()（确信）</span></span><br></pre></td></tr></table></figure>
还有ERCC基因一样的，选40%。</p>
<h4 id="筛选双细胞">筛选双细胞</h4>
<p>才发现这个要等到先聚类...别急。</p>
<h4 id="周期判断">周期判断</h4>
<p>这个要先标准化...别急...看<a target="_blank" rel="noopener" href="https://github.com/satijalab/seurat/issues/7942">seurat#7942</a>。
我说怎么把这两个放在后面。 <figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">db.test <span class="operator">&lt;-</span> scDblFinder<span class="operator">::</span>findDoubletClusters<span class="punctuation">(</span>GetAssayData<span class="punctuation">(</span>scRNA<span class="punctuation">,</span>slot<span class="operator">=</span><span class="string">"counts"</span><span class="punctuation">,</span>assay<span class="operator">=</span><span class="string">"RNA"</span><span class="punctuation">)</span><span class="punctuation">,</span> clusters<span class="operator">=</span>scRNA<span class="operator">@</span>meta.data<span class="operator">$</span>seurat_clusters<span class="punctuation">)</span></span><br></pre></td></tr></table></figure></p>
<h4 id="质控绘图">质控绘图</h4>
<p>最后就是绘图。<br>
看他们都是 - 图A：观察不同组cell的counts、feature分布 -
图B：nCount_RNA与对应的nFeature_RNA关系</p>
<p>图A用小提琴图，复习一下，VlnPlot()：</p>
<blockquote>
<ul>
<li><code>object</code>：<code>Seurat</code>对象。</li>
<li><code>features</code>：要绘制的特征。</li>
<li><code>group.by</code>：用于分组的元数据。</li>
<li><code>pt.size</code>：点的大小。</li>
<li><code>cols</code>：颜色。可以用<code>rainbow(n)</code>生成。</li>
</ul>
</blockquote>
<p>然后可以用<code>+</code>连接修改，theme(legend.position =
"none")隐藏图例，labs(tag = "A")添加全局标签A。</p>
<p>图B用FeatureScatter()，绘制两个特征之间的关系。</p>
<blockquote>
<ul>
<li><code>object</code>：<code>Seurat</code>对象。</li>
<li><code>feature1</code>：x轴特征。</li>
<li><code>feature2</code>：y轴特征。</li>
<li><code>group.by</code>：用于分组的元数据。</li>
<li>剩下通用的都差不多。</li>
</ul>
</blockquote>
<p>案例还对ERCC基因比例和nCount用了<code>FeatureScatter()</code>，这个图一看也验证了上面说的，ERCC基因百分比高，nCount就低。</p>
<h3 id="挑选高变异基因-1">挑选高变异基因</h3>
<p>用<code>FindVariableFeatures()</code>：</p>
<blockquote>
<ul>
<li><code>object</code>：<code>Seurat</code>对象。</li>
<li><code>selection.method</code>：选择方法。可以是"vst"（variance
stabilizing transformation）或"mean.var.plot"（均值方差图）。</li>
<li><code>nfeatures</code>：选择为顶部变异特征的特征数量；只在selection.method设置为'dispersion'或'vst'时使用。</li>
<li>还有一些其它的，还有使用vst法时的参数。</li>
</ul>
</blockquote>
<p>然后用<code>VariableFeaturePlot()</code>绘制变异特征图。<br>
<code>LabelPoints()</code>可以标记点，<code>head()</code>就可以取前几个。</p>
<h3 id="降维与聚类主要是找marker基因">降维与聚类，主要是找marker基因</h3>
<h4 id="pca">PCA</h4>
<p>要先标准化，这里用<code>ScaleData()</code>，对每个基因的表达量进行缩放和中心化，可以消除不同基因之间由于表达量级别不同而产生的差异。而不止用<code>NormalizeData()</code>，因为它对每个细胞之间的某个基因表达量进行归一化，可以消除不同细胞之间由于测序深度不同而产生的差异，所以归一化后，不同的基因可能仍然有不同的表达量级别。</p>
<p>然后就是简单的<code>scRNA &lt;- RunPCA(scRNA, features = VariableFeatures(scRNA))</code>，绘图可以用<code>DimPlot()</code>。</p>
<p>最后挑选主成分，<code>jackStraw()</code>评估每个基因的PCA得分的显著性（对于每个主成分，它会对数据进行PCA，然后对PCA的结果进行随机重排，然后再次进行PCA，然后计算原始PCA（实际）和重排PCA（随机下理论）的差异，这个差异越大，说明这个主成分越重要。），<code>ScoreJackStraw</code>用于分析jackStraw函数的结果，<code>JackStrawPlot()</code>绘制jackStraw图。</p>
这个图，empirical是实际的PCA得分，theoretical是随机重排的PCA得分（这两个对应彩色点），虚线是"均匀分布的
p
值的累积分布函数"，如果主成分轴显着，则二者应显示出差异。<del>不是很清楚，反正选不一样的就完了。</del><br>
这里结果都挺好的，盗用一下某人的例子，像11、12这种就不应该选了： <img src="https://cdn-ak.f.st-hatena.com/images/fotolife/k/kimoppy126/20180902/20180902233845.png">
<center>
<a target="_blank" rel="noopener" href="https://www.kimoton.com/entry/20180903/1535903928">《JackStraw
Plot (QQ-plot) を理解する》</a>
</center>
<h4 id="聚类还有一堆东西">聚类还有一堆东西</h4>
<p>烦了，调包就完了。</p>
<p>这里结果可能不太一样（<del>甚至教程的复现和原文不一样</del>），数据集应该是没变，重复运行一样应该也不是随机过程的问题，<del>暂且按下不表就当是版本差异</del>。</p>
<p><code>FindAllMarkers()</code>找marker基因，真慢啊，有必要换RRO或者python。<br>
完了再筛选出P&lt;0.05的，还有在不同类型细胞中表达量差异大的基因。</p>
<h3 id="qc之二">QC之二</h3>
<p>刚才说的要先聚类或标准化的筛选双细胞和周期判断。</p>
<p><code>emptyDrops()</code>检测空液滴，而检测双细胞可以用<code>doubletFinder()</code>，案例里的<code>doubletCluster()</code>其实寻找的是表达介于两个Cluster之间的Cluster，但他解释的是“有无double
droplet聚在一起的类”（而且好像已经废弃了）。<br>
<code>CellCycleScoring()</code>评估细胞周期。</p>
<h3 id="细胞类型注释-1">细胞类型注释</h3>
<p>按下不表。</p>
<h3 id="轨迹分析-1">轨迹分析</h3>
<p>选择特定基因代表细胞不同阶段的特征，然后用<code>Monocle</code>包，<code>newCellDataSet()</code>创建数据集，<code>estimateSizeFactors()</code>估计因子，<code>estimateDispersions()</code>估计离散度，<code>reduceDimension()</code>降维，<code>clusterCells()</code>聚类，<code>plot_cell_trajectory()</code>绘制轨迹图。</p>
<h2 id="总结">总结</h2>
<p>前面这一半的最终目标就是找出分化相关基因。<br>
抛开具体的实现来说，要做的事就是：</p>
<ul>
<li>质控：提升数据质量</li>
<li>降维：减少冗余信息</li>
<li>聚类：找出数据的内在结构</li>
<li>找marker基因</li>
</ul>
<p>还顺带做了轨迹分析利于解析分化的调控机制、找分化路径之类的。</p>
<p>有时间再用Python实现一下。
<font color="red"><strong>おわり</strong></font></p>

        
      </div>

         
    </div>
    
     
  </div>
  
    
<nav id="article-nav">
  <a class="article-nav-btn left "
    
      href="/blog/2024/07/29/legion-power-repaire/"
      title="拯救者电源接口接触不良维修"
     >
    <i class="fa-solid fa-angle-left"></i>
    <p class="title-text">
      
        拯救者电源接口接触不良维修
        
    </p>
  </a>
  <a class="article-nav-btn right "
    
      href="/blog/2024/04/11/mathmeth/"
      title="一些胡思乱想"
     >

    <p class="title-text">
      
        一些胡思乱想
        
    </p>
    <i class="fa-solid fa-angle-right"></i>
  </a>
</nav>


  
</article>




    </div>
    <div id="footer-wrapper">
      <footer id="footer">
  
  <div id="footer-info" class="inner">
    
    &copy; 2024 kusurin<br>
    Powered by <a href="https://hexo.io/" target="_blank">Hexo</a> & Theme <a target="_blank" rel="noopener" href="https://github.com/saicaca/hexo-theme-vivia">Vivia</a>
  </div>
</footer>

    </div>
    <div class="back-to-top-wrapper">
    <button id="back-to-top-btn" class="back-to-top-btn hide" onclick="topFunction()">
        <i class="fa-solid fa-angle-up"></i>
    </button>
</div>

<script>
    function topFunction() {
        window.scroll({ top: 0, behavior: 'smooth' });
    }
    let btn = document.getElementById('back-to-top-btn');
    function scrollFunction() {
        if (document.body.scrollTop > 600 || document.documentElement.scrollTop > 600) {
            btn.classList.remove('hide')
        } else {
            btn.classList.add('hide')
        }
    }
    window.onscroll = function() {
        scrollFunction();
    }
</script>

  </div>
  <script src="/blog/js/light-dark-switch.js"></script>
<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        tex2jax: {
            inlineMath: [ ["$","$"], ["\\(","\\)"] ],
            skipTags: ['script', 'noscript', 'style', 'textarea', 'pre', 'code'],
            processEscapes: true
        }
    });
    MathJax.Hub.Queue(function() {
        var all = MathJax.Hub.getAllJax();
        for (var i = 0; i < all.length; ++i)
            all[i].SourceElement().parentNode.className += ' has-jax';
    });
</script>
<script src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
</body>
</html>
