<!DOCTYPE html>


<html theme="dark" showBanner="true" hasBanner="false" > 
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free@6.5.1/css/all.min.css" integrity="sha256-wiz7ZSCn/btzhjKDQBms9Hx4sSeUYsDrTLg7roPstac=" crossorigin="anonymous">
<script src="/blog/js/color.global.min.js" ></script>
<script src="/blog/js/load-settings.js" ></script>

<!-- Google tag (gtag.js) -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-31GM7RWSBJ"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-31GM7RWSBJ');
</script>

<head>
  <meta charset="utf-8">
  
  
  

  
  <title>单细胞数据挖掘 | kusurin.blog</title>
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">

  <meta name="description" content="案例里用了《Glioblastoma cell differentiation trajectory predicts the immunotherapy response and overall survival of patients》，主要是使用scRNA-seq（单细胞RNA测序）和bulk RNA-seq（整体RNA测序）来找出GBM分化相关基因然后预测患者总生存期之类的。 实验">
<meta property="og:type" content="article">
<meta property="og:title" content="单细胞数据挖掘">
<meta property="og:url" content="https://kusurin.github.io/blog/2024/05/22/single-cell-data-mining/index.html">
<meta property="og:site_name" content="kusurin.blog">
<meta property="og:description" content="案例里用了《Glioblastoma cell differentiation trajectory predicts the immunotherapy response and overall survival of patients》，主要是使用scRNA-seq（单细胞RNA测序）和bulk RNA-seq（整体RNA测序）来找出GBM分化相关基因然后预测患者总生存期之类的。 实验">
<meta property="og:locale" content="zh_CN">
<meta property="og:image" content="https://kusurin.github.io/blog/images/square_hole.png">
<meta property="article:published_time" content="2024-05-22T15:14:26.000Z">
<meta property="article:modified_time" content="2025-03-13T12:45:39.231Z">
<meta property="article:author" content="kusurin">
<meta property="article:tag" content="学习笔记">
<meta property="article:tag" content="生信">
<meta property="article:tag" content="scRNA-seq">
<meta property="article:tag" content="R">
<meta property="article:tag" content="Python">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://kusurin.github.io/blog/images/square_hole.png">
  
    <link rel="alternate" href="/blog/atom.xml" title="kusurin.blog" type="application/atom+xml">
  
  
    <link rel="icon" media="(prefers-color-scheme: light)" href="/blog/images/favicon-light-32.png" sizes="32x32">
    <link rel="icon" media="(prefers-color-scheme: light)" href="/blog/images/favicon-light-128.png" sizes="128x128">
    <link rel="icon" media="(prefers-color-scheme: light)" href="/blog/images/favicon-light-180.png" sizes="180x180">
    <link rel="icon" media="(prefers-color-scheme: light)" href="/blog/images/favicon-light-192.png" sizes="192x192">
    <link rel="icon" media="(prefers-color-scheme: dark)" href="/blog/images/favicon-dark-32.png" sizes="32x32">
    <link rel="icon" media="(prefers-color-scheme: dark)" href="/blog/images/favicon-dark-128.png" sizes="128x128">
    <link rel="icon" media="(prefers-color-scheme: dark)" href="/blog/images/favicon-dark-180.png" sizes="180x180">
    <link rel="icon" media="(prefers-color-scheme: dark)" href="/blog/images/favicon-dark-192.png" sizes="192x192">
  
  
<link rel="stylesheet" href="/blog/css/style.css">

<meta name="generator" content="Hexo 6.3.0"></head>


<body>
  
   
  <div id="main-grid" class="shadow   ">
    <div id="nav" class=""  >
      <navbar id="navbar">
  <nav id="title-nav">
    <a href="/blog/">
      <div id="vivia-logo">
        <div class="dot"></div>
        <div class="dot"></div>
        <div class="dot"></div>
        <div class="dot"></div>
      </div>
      <div>kusurin.blog </div>
    </a>
  </nav>
  <nav id="main-nav">
    
      <a class="main-nav-link" href="/blog/">Home</a>
    
      <a class="main-nav-link" href="/blog/archives">Archives</a>
    
      <a class="main-nav-link" href="/blog/about">About</a>
    
  </nav>
  <nav id="sub-nav">
    <a id="theme-btn" class="nav-icon">
      <span class="light-mode-icon"><svg xmlns="http://www.w3.org/2000/svg" height="20" viewBox="0 -960 960 960" width="20"><path d="M438.5-829.913v-48q0-17.452 11.963-29.476 11.964-12.024 29.326-12.024 17.363 0 29.537 12.024t12.174 29.476v48q0 17.452-11.963 29.476-11.964 12.024-29.326 12.024-17.363 0-29.537-12.024T438.5-829.913Zm0 747.826v-48q0-17.452 11.963-29.476 11.964-12.024 29.326-12.024 17.363 0 29.537 12.024t12.174 29.476v48q0 17.452-11.963 29.476-11.964 12.024-29.326 12.024-17.363 0-29.537-12.024T438.5-82.087ZM877.913-438.5h-48q-17.452 0-29.476-11.963-12.024-11.964-12.024-29.326 0-17.363 12.024-29.537t29.476-12.174h48q17.452 0 29.476 11.963 12.024 11.964 12.024 29.326 0 17.363-12.024 29.537T877.913-438.5Zm-747.826 0h-48q-17.452 0-29.476-11.963-12.024-11.964-12.024-29.326 0-17.363 12.024-29.537T82.087-521.5h48q17.452 0 29.476 11.963 12.024 11.964 12.024 29.326 0 17.363-12.024 29.537T130.087-438.5Zm660.174-290.87-34.239 32q-12.913 12.674-29.565 12.174-16.653-.5-29.327-13.174-12.674-12.673-12.554-28.826.12-16.152 12.794-28.826l33-35q12.913-12.674 30.454-12.674t30.163 12.847q12.709 12.846 12.328 30.826-.38 17.98-13.054 30.653ZM262.63-203.978l-32 34q-12.913 12.674-30.454 12.674t-30.163-12.847q-12.709-12.846-12.328-30.826.38-17.98 13.054-30.653l33.239-31q12.913-12.674 29.565-12.174 16.653.5 29.327 13.174 12.674 12.673 12.554 28.826-.12 16.152-12.794 28.826Zm466.74 33.239-32-33.239q-12.674-12.913-12.174-29.565.5-16.653 13.174-29.327 12.673-12.674 28.826-13.054 16.152-.38 28.826 12.294l35 33q12.674 12.913 12.674 30.454t-12.847 30.163q-12.846 12.709-30.826 12.328-17.98-.38-30.653-13.054ZM203.978-697.37l-34-33q-12.674-12.913-13.174-29.945-.5-17.033 12.174-29.707t31.326-13.293q18.653-.62 31.326 13.054l32 34.239q11.674 12.913 11.174 29.565-.5 16.653-13.174 29.327-12.673 12.674-28.826 12.554-16.152-.12-28.826-12.794ZM480-240q-100 0-170-70t-70-170q0-100 70-170t170-70q100 0 170 70t70 170q0 100-70 170t-170 70Zm-.247-82q65.703 0 111.475-46.272Q637-414.544 637-480.247t-45.525-111.228Q545.95-637 480.247-637t-111.475 45.525Q323-545.95 323-480.247t45.525 111.975Q414.05-322 479.753-322ZM481-481Z"/></svg></span>
      <span class="dark-mode-icon"><svg xmlns="http://www.w3.org/2000/svg" height="20" viewBox="0 -960 960 960" width="20"><path d="M480.239-116.413q-152.63 0-258.228-105.478Q116.413-327.37 116.413-480q0-130.935 77.739-227.435t206.304-125.043q43.022-9.631 63.87 10.869t3.478 62.805q-8.891 22.043-14.315 44.463-5.424 22.42-5.424 46.689 0 91.694 64.326 155.879 64.325 64.186 156.218 64.186 24.369 0 46.978-4.946 22.609-4.945 44.413-14.076 42.826-17.369 62.967 1.142 20.142 18.511 10.511 61.054Q807.174-280 712.63-198.206q-94.543 81.793-232.391 81.793Zm0-95q79.783 0 143.337-40.217 63.554-40.218 95.793-108.283-15.608 4.044-31.097 5.326-15.49 1.283-31.859.805-123.706-4.066-210.777-90.539-87.071-86.473-91.614-212.092-.24-16.369.923-31.978 1.164-15.609 5.446-30.978-67.826 32.478-108.282 96.152Q211.652-559.543 211.652-480q0 111.929 78.329 190.258 78.329 78.329 190.258 78.329ZM466.13-465.891Z"/></svg></span>
    </a>
    
      <a id="nav-rss-link" class="nav-icon mobile-hide" href="/blog/atom.xml" title="RSS 订阅">
        <svg xmlns="http://www.w3.org/2000/svg" height="20" viewBox="0 -960 960 960" width="20"><path d="M198-120q-25.846 0-44.23-18.384-18.384-18.385-18.384-44.23 0-25.846 18.384-44.23 18.384-18.385 44.23-18.385 25.846 0 44.23 18.385 18.384 18.384 18.384 44.23 0 25.845-18.384 44.23Q223.846-120 198-120Zm538.385 0q-18.846 0-32.923-13.769-14.076-13.769-15.922-33.23-8.692-100.616-51.077-188.654-42.385-88.039-109.885-155.539-67.5-67.501-155.539-109.885Q283-663.462 182.385-672.154q-19.461-1.846-33.23-16.23-13.769-14.385-13.769-33.846t14.076-32.922q14.077-13.461 32.923-12.23 120.076 8.692 226.038 58.768 105.961 50.077 185.73 129.846 79.769 79.769 129.846 185.731 50.077 105.961 58.769 226.038 1.231 18.846-12.538 32.922Q756.461-120 736.385-120Zm-252 0q-18.231 0-32.423-13.461t-18.653-33.538Q418.155-264.23 348.886-333.5q-69.27-69.27-166.501-84.423-20.077-4.462-33.538-18.961-13.461-14.5-13.461-33.346 0-19.076 13.884-33.23 13.884-14.153 33.115-10.922 136.769 15.384 234.384 112.999 97.615 97.615 112.999 234.384 3.231 19.23-10.538 33.115Q505.461-120 484.385-120Z"/></svg>
      </a>
    
    <div id="nav-menu-btn" class="nav-icon">
      <svg xmlns="http://www.w3.org/2000/svg" height="20" viewBox="0 -960 960 960" width="20"><path d="M177.37-252.282q-17.453 0-29.477-11.964-12.024-11.963-12.024-29.326t12.024-29.537q12.024-12.174 29.477-12.174h605.26q17.453 0 29.477 11.964 12.024 11.963 12.024 29.326t-12.024 29.537q-12.024 12.174-29.477 12.174H177.37Zm0-186.218q-17.453 0-29.477-11.963-12.024-11.964-12.024-29.326 0-17.363 12.024-29.537T177.37-521.5h605.26q17.453 0 29.477 11.963 12.024 11.964 12.024 29.326 0 17.363-12.024 29.537T782.63-438.5H177.37Zm0-186.217q-17.453 0-29.477-11.964-12.024-11.963-12.024-29.326t12.024-29.537q12.024-12.174 29.477-12.174h605.26q17.453 0 29.477 11.964 12.024 11.963 12.024 29.326t-12.024 29.537q-12.024 12.174-29.477 12.174H177.37Z"/></svg>
    </div>
  </nav>
</navbar>
<div id="nav-dropdown" class="hidden">
  <div id="dropdown-link-list">
    
      <a class="nav-dropdown-link" href="/blog/">Home</a>
    
      <a class="nav-dropdown-link" href="/blog/archives">Archives</a>
    
      <a class="nav-dropdown-link" href="/blog/about">About</a>
    
    
      <a class="nav-dropdown-link" href="/blog/atom.xml" title="RSS 订阅">RSS</a>
     
    </div>
</div>
<script>
  let dropdownBtn = document.getElementById("nav-menu-btn");
  let dropdownEle = document.getElementById("nav-dropdown");
  dropdownBtn.onclick = function() {
    dropdownEle.classList.toggle("hidden");
  }
</script>
    </div>
    <div id="sidebar-wrapper">
      <sidebar id="sidebar">
  
    <div class="widget-wrap">
  <div class="info-card">
    <div class="avatar">
      
        <image src=https://pic1.imgdb.cn/item/6828010758cb8da5c8f7d65d.gif></image>
      
      <div class="img-dim"></div>
    </div>
    <div class="info">
      <div class="username">kusurin </div>
      <div class="dot"></div>
      <div class="subtitle">文字の中に沈む </div>
      <div class="link-list">
        
          <a class="link-btn" target="_blank" rel="noopener" href="https://kusurin.icu" title="Home Page"><i class="fa-solid fa-globe"></i></a>
        
          <a class="link-btn" target="_blank" rel="noopener" href="https://kusurin.icu/blog" title="Blog Home Page"><i class="fa-solid fa-house"></i></a>
        
          <a class="link-btn" target="_blank" rel="noopener" href="https://github.com/kusurin" title="GitHub"><i class="fa-brands fa-github"></i></a>
        
          <a class="link-btn" target="_blank" rel="noopener" href="https://space.bilibili.com/5404263" title="bilibili"><i class="fa-brands fa-bilibili"></i></a>
         
      </div>  
    </div>
  </div>
</div>

  
  <div class="sticky">
    
      


  <div class="widget-wrap">
    <div class="widget">
      <h3 class="widget-title">分类</h3>
      <div class="category-box">
            <a class="category-link" href="/blog/categories/Coding/">
                Coding
                <div class="category-count">2</div>
            </a>
        
            <a class="category-link" href="/blog/categories/%E6%9D%82%E8%B0%88/">
                杂谈
                <div class="category-count">13</div>
            </a>
        
            <a class="category-link" href="/blog/categories/CSS-tricks/">
                CSS-tricks
                <div class="category-count">3</div>
            </a>
        
            <a class="category-link" href="/blog/categories/%E8%B8%A9%E5%9D%91%E6%97%A5%E8%AE%B0/">
                踩坑日记
                <div class="category-count">5</div>
            </a>
        
            <a class="category-link" href="/blog/categories/%E7%94%9F%E4%BF%A1/">
                生信
                <div class="category-count">8</div>
            </a>
        </div>
    </div>
  </div>


    
      
  <div class="widget-wrap">
    <div class="widget">
      <h3 class="widget-title">标签</h3>
      <ul class="widget-tag-list" itemprop="keywords"><li class="widget-tag-list-item"><a class="widget-tag-list-link" href="/blog/tags/CSS/" rel="tag">CSS</a></li><li class="widget-tag-list-item"><a class="widget-tag-list-link" href="/blog/tags/Linux/" rel="tag">Linux</a></li><li class="widget-tag-list-item"><a class="widget-tag-list-link" href="/blog/tags/Python/" rel="tag">Python</a></li><li class="widget-tag-list-item"><a class="widget-tag-list-link" href="/blog/tags/R/" rel="tag">R</a></li><li class="widget-tag-list-item"><a class="widget-tag-list-link" href="/blog/tags/R-based/" rel="tag">R based</a></li><li class="widget-tag-list-item"><a class="widget-tag-list-link" href="/blog/tags/WPA2/" rel="tag">WPA2</a></li><li class="widget-tag-list-item"><a class="widget-tag-list-link" href="/blog/tags/scATAC-seq/" rel="tag">scATAC-seq</a></li><li class="widget-tag-list-item"><a class="widget-tag-list-link" href="/blog/tags/scRNA-seq/" rel="tag">scRNA-seq</a></li><li class="widget-tag-list-item"><a class="widget-tag-list-link" href="/blog/tags/%E5%82%85%E7%AB%8B%E5%8F%B6/" rel="tag">傅立叶</a></li><li class="widget-tag-list-item"><a class="widget-tag-list-link" href="/blog/tags/%E5%AD%A6%E4%B9%A0%E7%AC%94%E8%AE%B0/" rel="tag">学习笔记</a></li><li class="widget-tag-list-item"><a class="widget-tag-list-link" href="/blog/tags/%E6%9C%BA%E5%99%A8%E5%AD%A6%E4%B9%A0/" rel="tag">机器学习</a></li><li class="widget-tag-list-item"><a class="widget-tag-list-link" href="/blog/tags/%E6%9D%82%E8%B0%88/" rel="tag">杂谈</a></li><li class="widget-tag-list-item"><a class="widget-tag-list-link" href="/blog/tags/%E6%A8%A1%E6%8B%9F%E4%BD%8D%E7%BD%AE/" rel="tag">模拟位置</a></li><li class="widget-tag-list-item"><a class="widget-tag-list-link" href="/blog/tags/%E7%94%9F%E4%BF%A1/" rel="tag">生信</a></li><li class="widget-tag-list-item"><a class="widget-tag-list-link" href="/blog/tags/%E7%A9%BA%E9%97%B4%E8%BD%AC%E5%BD%95%E7%BB%84/" rel="tag">空间转录组</a></li><li class="widget-tag-list-item"><a class="widget-tag-list-link" href="/blog/tags/%E7%AE%97%E6%B3%95/" rel="tag">算法</a></li><li class="widget-tag-list-item"><a class="widget-tag-list-link" href="/blog/tags/%E7%BB%B4%E4%BF%AE/" rel="tag">维修</a></li><li class="widget-tag-list-item"><a class="widget-tag-list-link" href="/blog/tags/%E7%BD%91%E5%AE%89/" rel="tag">网安</a></li><li class="widget-tag-list-item"><a class="widget-tag-list-link" href="/blog/tags/%E8%B8%A9%E5%9D%91/" rel="tag">踩坑</a></li><li class="widget-tag-list-item"><a class="widget-tag-list-link" href="/blog/tags/%E8%BE%B9%E6%A1%86/" rel="tag">边框</a></li></ul>
    </div>
  </div>


    
  </div>
</sidebar>
    </div>
    <div id="content-body">
       


<article id="post-single-cell-data-mining" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  
    
   
  <div class="article-inner">
    <div class="article-main">
      <header class="article-header">
        
<div class="main-title-bar">
  <div class="main-title-dot"></div>
  
    
      <h1 class="p-name article-title" itemprop="headline name">
        单细胞数据挖掘
      </h1>
    
  
</div>

        <div class='meta-info-bar'>
          <div class="meta-info">
  <time class="dt-published" datetime="2024-05-22T15:14:26.000Z" itemprop="datePublished">2024-05-22</time>
</div>
          <div class="need-seperator meta-info">
            <div class="meta-cate-flex">
  
  <a class="meta-cate-link" href="/blog/categories/%E7%94%9F%E4%BF%A1/">生信</a>
   
</div>
  
          </div>
          <div class="wordcount need-seperator meta-info">
            10k 词 
          </div>
        </div>
        
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/blog/tags/Python/" rel="tag">Python</a></li><li class="article-tag-list-item"><a class="article-tag-list-link" href="/blog/tags/R/" rel="tag">R</a></li><li class="article-tag-list-item"><a class="article-tag-list-link" href="/blog/tags/scRNA-seq/" rel="tag">scRNA-seq</a></li><li class="article-tag-list-item"><a class="article-tag-list-link" href="/blog/tags/%E5%AD%A6%E4%B9%A0%E7%AC%94%E8%AE%B0/" rel="tag">学习笔记</a></li><li class="article-tag-list-item"><a class="article-tag-list-link" href="/blog/tags/%E7%94%9F%E4%BF%A1/" rel="tag">生信</a></li></ul>

      </header>
      <div class="e-content article-entry" itemprop="articleBody">
        
          
    <div id="toc">
        <strong class="sidebar-title">目录</strong>
        <ol class="toc"><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%AE%9E%E9%AA%8C%E5%8E%9F%E7%90%86"><span class="toc-text">实验原理</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#%E6%95%B0%E6%8D%AE%E9%9B%86"><span class="toc-text">数据集</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E8%B4%A8%E6%8E%A7"><span class="toc-text">质控</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E6%8C%91%E9%80%89%E9%AB%98%E5%8F%98%E5%BC%82%E5%9F%BA%E5%9B%A0"><span class="toc-text">挑选高变异基因</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E9%99%8D%E7%BB%B4%E4%B8%8E%E8%81%9A%E7%B1%BB"><span class="toc-text">降维与聚类</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E7%BB%86%E8%83%9E%E7%B1%BB%E5%9E%8B%E6%B3%A8%E9%87%8A"><span class="toc-text">细胞类型注释</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E8%BD%A8%E8%BF%B9%E5%88%86%E6%9E%90"><span class="toc-text">轨迹分析</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#r%E8%AF%AD%E8%A8%80%E5%AE%9E%E7%8E%B0"><span class="toc-text">R语言实现</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#%E6%8F%90%E5%8F%96%E6%95%B0%E6%8D%AE"><span class="toc-text">提取数据</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E8%B4%A8%E6%8E%A7-1"><span class="toc-text">质控</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E6%8C%91%E9%80%89%E9%AB%98%E5%8F%98%E5%BC%82%E5%9F%BA%E5%9B%A0-1"><span class="toc-text">挑选高变异基因</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E9%99%8D%E7%BB%B4%E4%B8%8E%E8%81%9A%E7%B1%BB%E4%B8%BB%E8%A6%81%E6%98%AF%E6%89%BEmarker%E5%9F%BA%E5%9B%A0"><span class="toc-text">降维与聚类，主要是找marker基因</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#qc%E4%B9%8B%E4%BA%8C"><span class="toc-text">QC之二</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E7%BB%86%E8%83%9E%E7%B1%BB%E5%9E%8B%E6%B3%A8%E9%87%8A-1"><span class="toc-text">细胞类型注释</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E8%BD%A8%E8%BF%B9%E5%88%86%E6%9E%90-1"><span class="toc-text">轨迹分析</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%B0%BE%E5%A3%B0"><span class="toc-text">尾声</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#%E5%A4%9A%E6%A0%B7%E6%9C%AC%E6%95%B4%E5%90%88"><span class="toc-text">多样本整合</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#infecnv"><span class="toc-text">infeCNV</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#copykat"><span class="toc-text">copyKAT</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E7%BB%86%E8%83%9E%E9%80%9A%E8%AE%AF"><span class="toc-text">细胞通讯</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#python%E5%AE%9E%E7%8E%B0"><span class="toc-text">Python实现</span></a></li></ol>
    </div>

          <p>案例里用了《Glioblastoma cell differentiation trajectory predicts the
immunotherapy response and overall survival of
patients》，主要是使用scRNA-seq（单细胞RNA测序）和bulk
RNA-seq（整体RNA测序）来找出GBM分化相关基因然后预测患者总生存期之类的。</p>
<h2 id="实验原理">实验原理</h2>
<p>主要就是干了质控、基因可视化、降维与聚类、周期判断、double
droplet、细胞类型注释、拟时序分析。</p>
<h3 id="数据集">数据集</h3>
<p>这堆数据集就是一轴是细胞，一轴是基因种类，每个点代表某个细胞/样本的某个基因表达量。</p>
<h3 id="质控">质控</h3>
<p>质控主要就是过滤：</p>
<ul>
<li>低质量的细胞(比如基因少、线粒体基因多)</li>
<li>双细胞（两个细胞被捕获到同一个液滴中）</li>
<li>高ERCC基因（External RNA Controls
Consortium），<em>“是一组合成的RNA分子，它们的浓度已知，并且在测序实验中添加到样本中，以便可以用来校正技术变异和进行数据标准化”</em>，在所有细胞中的浓度应该是一致的，所以如果ERCC基因百分比高，那这个细胞的RNA可能较少。</li>
</ul>
<p><code>Seurat</code>包可以傻瓜实现。</p>
<p>案例里还提到了double droplet和周期判断。<br />
double droplet就是上面说的双细胞。<br />
而周期判断就是说，一些细胞处于分裂期，它们会高表达某些基因，可能会影响聚类，所以要去掉。<br />
<code>scran</code>包可以傻瓜实现。</p>
<h3 id="挑选高变异基因">挑选高变异基因</h3>
<p>highly Variable
gene是在一组样本或细胞中，表达水平变化大的基因，它们可能是驱动细胞异质性的关键。
<code>Seurat</code>包同样可以傻瓜实现。</p>
<h3 id="降维与聚类">降维与聚类</h3>
<p>降维机器学习里也有，<em>“将数据从原始的高维空间转换到低维空间，同时尽可能保留数据的变异性”</em>，总的来说目的就是去掉冗杂信息。一个方法就是PCA（主成分分析），能使得降维后数据的方差尽可能大（保留变异性），而和原本数据的均方误差尽可能小(减小和原数据的差距)，这两个其实殊途同归。更多可见<a
href="/blog/2025/01/21/ML-induction/#降维">另一篇博客</a>。<br />
PCA对数据的尺度（可以用标准差度量）非常敏感，如果不同的特征有不同的尺度（0-1vs0-1000），那么PCA的结果将主要受到尺度大的特征的影响，所以PCA之前可以先缩放特征到同一尺度。</p>
<p>聚类就是把数据分成几类，使得同一个类内数据的相似性尽可能大，同时不同类数据的差异性也尽可能地大。聚类是一种无监督学习，也就是机器自己把数据根据特征分类，而不需要标注的指导。更多可见<a
href="/blog/2025/01/21/ML-induction/#聚类">另一篇博客</a>。</p>
<p>案例里PCA之后聚类然后又用t-SNE降维（也可以用UMAP），后者主要是为了降至2或3维方便可视化。</p>
<style>
    #square-hole img {
        display:none;
        position:absolute;
        left:0;
        top:100%;
    }
    #square-hole:hover img {
        display:block;
    }
</style>
<p><code>Seurat</code>包还是可以傻瓜实现。<a id="square-hole" onfocus="" ontouchstart="" style="position:relative;display:inline-block;word-break:all;"><del>that's
right the square hole</del>
<img src="/blog/images/square_hole.png" style=""></a></p>
<h3 id="细胞类型注释">细胞类型注释</h3>
<p>字面意思，就比如把一些细胞标注成B细胞之类的甚至更精细一些，幼稚B细胞之类的。这靠的是一些特异性高的marker基因。
也可以用<code>SingleR</code>包自动完成（不推荐）。</p>
<h3 id="轨迹分析">轨迹分析</h3>
<p>根据细胞的表达谱，找出连续变化的基因指示细胞的变化轨迹，推断出一种可能的细胞状态随时间的变化过程，要注意起点和终点并不一定对应着真正时间上的起点和终点。<br />
可以用<code>Monocle</code>包，3代比2代比起来，可以人工指定起点和终点。</p>
<hr />
<p><span
style="color:#808080;"><em><del>本来这点东西一天下班时间就能搞完的，现在下来两三天了，不光是因为只有业余时间，还真是怠惰啊。</del></em></span></p>
<hr />
<h2 id="r语言实现">R语言实现</h2>
<p>代码详见<a
target="_blank" rel="noopener" href="https://github.com/kusurin/single-cell-data-mining">single-cell-data-mining</a>。</p>
<h3 id="提取数据">提取数据</h3>
<p>怎么两个数据集形式差那么多，你们能不能统一一下（骂骂咧咧）。</p>
<p>先看GSE84465学习一下。<br />
案例里的GSE84465集合到了一个<code>.csv</code>里，列名是某样本的plate_id和Well，行名就是基因。<br />
然后还有一个SraRunTable.txt，刚刚还奇怪，上面的文件都没有分组信息，这个文件里就有，就是每个样本的一些元信息，比如组织类型、患者编号、对应的样本编号等。这个就是SRA里的元数据，进SRA
Run Selector里就能找到。<br />
所以这个SraRunTable.txt就是用来给上面那些scRNA-seq数据分组。</p>
<p>我们要做的就是从SRA生成GSE对应的列名，并使用SRA的信息筛选出肿瘤组织，借助对应的列名筛选出GSE中的肿瘤组织。<br />
还有SRA记录着分组信息，也要一起筛选出来。</p>
<p>所以说拿到GSE112302也是一样的，就是GSE112302没有给全部合在一起的数据，需要我们自己合起来（附加列，也就是样本），然后用SRA筛选出肿瘤组织。</p>
<h3 id="质控-1">质控</h3>
<h4
id="创建seurat对象简单筛选细胞">创建<code>Seurat</code>对象，简单筛选细胞</h4>
<p>要先创建<code>Seurat</code>对象。<code>CreateSeuratObject()</code>函数的文档如下：</p>
<blockquote>
<ul>
<li><code>counts</code>：这是一个矩阵类对象，包含未经标准化的数据，其中细胞作为列，特征作为行，或者是一个<code>Assay</code>派生对象。</li>
<li><code>assay</code>：初始count矩阵的名称。</li>
<li><code>meta.data</code>：要添加到<code>Seurat</code>对象的额外细胞级元数据。应该是一个数据框，其中行是细胞名称，列是额外的元数据字段。元数据中的行名称需要与计数矩阵的列名称匹配。</li>
<li><code>project</code>：<code>Seurat</code>对象的项目名称。</li>
<li><code>...</code>：传递给其他方法的参数。</li>
<li><code>min.cells</code>：至少在这么多细胞中检测到的特征。也会对计数矩阵进行取子集。要重新引入排除的特征，使用较低的截止值创建一个新对象。</li>
<li><code>min.features</code>：至少在这么多特征中检测到的细胞。</li>
</ul>
</blockquote>
<p>（对了R语言的函数是这样调用的：<code>function_name(arg1 = value1, arg2 = value2, ...)</code>）</p>
<p>其实只要用到<code>counts</code>、<code>meta.data</code>、<code>min.cells</code>、<code>min.features</code>就行了。
这里<code>counts</code>就是count矩阵，<code>meta.data</code>就是分组信息，用<code>data.frame()</code>上面说了是这样的：<code>...</code>是列（额外的元数据字段，这里就是病人编号，要用来分组嘛），<code>row.names</code>是行（样本编号）,要和<code>counts</code>的列名对应。<br />
然后后两个前者用来筛选表达的细胞的数量低的基因，后者用来筛选基因少的细胞，顺便还能过滤空细胞。</p>
<h4
id="筛选线粒体基因和ercc基因过多细胞">筛选线粒体基因和ERCC基因过多细胞</h4>
<p>然后还要筛选掉线粒体基因过多的细胞： <figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">scRNA<span class="punctuation">[[</span><span class="string">&quot;percent.mt&quot;</span><span class="punctuation">]</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> PercentageFeatureSet<span class="punctuation">(</span>scRNA<span class="punctuation">,</span> pattern <span class="operator">=</span> <span class="string">&quot;^MT-&quot;</span><span class="punctuation">)</span> <span class="comment">#先全部算了比较好，PercentageFeatureSet()计算满足给定模式的特征在每个细胞中的百分比，后面那个参数就是正则（真是先进啊（大嘘））</span></span><br><span class="line">scRNA <span class="operator">&lt;-</span> subset<span class="punctuation">(</span>scRNA<span class="punctuation">,</span> subset <span class="operator">=</span> percent.mt <span class="operator">&lt;</span> <span class="number">5</span><span class="punctuation">)</span> <span class="comment">#案例取5%，subset()就是Array.filter()（确信）</span></span><br></pre></td></tr></table></figure>
还有ERCC基因一样的，选40%。</p>
<h4 id="筛选双细胞">筛选双细胞</h4>
<p>筛选双细胞应当在一个样本内进行，先一套下来到聚类，然后
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">sweep.res.list <span class="operator">&lt;-</span> paramSweep<span class="punctuation">(</span>seurat_object<span class="punctuation">,</span> PCs <span class="operator">=</span> pc.num<span class="punctuation">,</span> sct <span class="operator">=</span> <span class="built_in">F</span><span class="punctuation">)</span></span><br><span class="line">sweep.stats <span class="operator">&lt;-</span> summarizeSweep<span class="punctuation">(</span>sweep.res.list<span class="punctuation">,</span> GT <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span></span><br><span class="line">bcmvn <span class="operator">&lt;-</span> find.pK<span class="punctuation">(</span>sweep.stats<span class="punctuation">)</span></span><br><span class="line">pK_bcmvn <span class="operator">&lt;-</span> bcmvn<span class="operator">$</span>pK<span class="punctuation">[</span>which.max<span class="punctuation">(</span>bcmvn<span class="operator">$</span>BCmetric<span class="punctuation">)</span><span class="punctuation">]</span> <span class="operator">%&gt;%</span> <span class="built_in">as.character</span><span class="punctuation">(</span><span class="punctuation">)</span> <span class="operator">%&gt;%</span> <span class="built_in">as.numeric</span><span class="punctuation">(</span><span class="punctuation">)</span></span><br><span class="line">homotypic.prop <span class="operator">&lt;-</span> modelHomotypic<span class="punctuation">(</span>seurat_object<span class="operator">$</span>seurat_clusters<span class="punctuation">)</span></span><br><span class="line">nExp_poi <span class="operator">&lt;-</span> <span class="built_in">round</span><span class="punctuation">(</span>doublet.rate<span class="operator">*</span>ncol<span class="punctuation">(</span>seurat_object<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">nExp_poi.adj <span class="operator">&lt;-</span> <span class="built_in">round</span><span class="punctuation">(</span>nExp_poi<span class="operator">*</span><span class="punctuation">(</span><span class="number">1</span><span class="operator">-</span>homotypic.prop<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">seurat_object <span class="operator">&lt;-</span> doubletFinder<span class="punctuation">(</span>seurat_object<span class="punctuation">,</span> PCs <span class="operator">=</span> pc.num<span class="punctuation">,</span> pN <span class="operator">=</span> <span class="number">0.25</span><span class="punctuation">,</span> pK <span class="operator">=</span> pK_bcmvn<span class="punctuation">,</span> nExp <span class="operator">=</span> nExp_poi.adj<span class="punctuation">,</span> reuse.pANN <span class="operator">=</span> <span class="built_in">F</span><span class="punctuation">,</span> sct <span class="operator">=</span> <span class="built_in">F</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure></p>
<h4 id="周期分析">周期分析</h4>
<p>这个要先标准化...别急...看<a
target="_blank" rel="noopener" href="https://github.com/satijalab/seurat/issues/7942">seurat#7942</a>。
我说怎么把这两个放在后面。 <figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">seurat_object <span class="operator">&lt;-</span> CellCycleScoring<span class="punctuation">(</span>seurat_object<span class="punctuation">,</span> s.features<span class="operator">=</span> gene_s<span class="punctuation">,</span> g2m.features<span class="operator">=</span> gene_g2m<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>
<code>gene_s</code>和<code>gene_g2m</code>可以下载也可以自己提供，注意物种区别。</p>
<h4 id="质控绘图">质控绘图</h4>
<p>最后就是绘图。<br />
看他们都是 - 图A：观察不同组cell的counts、feature分布 -
图B：nCount_RNA与对应的nFeature_RNA关系</p>
<p>图A用小提琴图，复习一下，VlnPlot()：</p>
<blockquote>
<ul>
<li><code>object</code>：<code>Seurat</code>对象。</li>
<li><code>features</code>：要绘制的特征。</li>
<li><code>group.by</code>：用于分组的元数据。</li>
<li><code>pt.size</code>：点的大小。</li>
<li><code>cols</code>：颜色。可以用<code>rainbow(n)</code>生成。</li>
</ul>
</blockquote>
<p>然后可以用<code>+</code>连接修改，theme(legend.position =
"none")隐藏图例，labs(tag = "A")添加全局标签A。</p>
<p>图B用FeatureScatter()，绘制两个特征之间的关系。</p>
<blockquote>
<ul>
<li><code>object</code>：<code>Seurat</code>对象。</li>
<li><code>feature1</code>：x轴特征。</li>
<li><code>feature2</code>：y轴特征。</li>
<li><code>group.by</code>：用于分组的元数据。</li>
<li>剩下通用的都差不多。</li>
</ul>
</blockquote>
<p>案例还对ERCC基因比例和nCount用了<code>FeatureScatter()</code>，这个图一看也验证了上面说的，ERCC基因百分比高，nCount就低。</p>
<h3 id="挑选高变异基因-1">挑选高变异基因</h3>
<p>用<code>FindVariableFeatures()</code>：</p>
<blockquote>
<ul>
<li><code>object</code>：<code>Seurat</code>对象。</li>
<li><code>selection.method</code>：选择方法。可以是"vst"（variance
stabilizing transformation）或"mean.var.plot"（均值方差图）。</li>
<li><code>nfeatures</code>：选择为顶部变异特征的特征数量；只在selection.method设置为'dispersion'或'vst'时使用。</li>
<li>还有一些其它的，还有使用vst法时的参数。</li>
</ul>
</blockquote>
<p>然后用<code>VariableFeaturePlot()</code>绘制变异特征图。<br />
<code>LabelPoints()</code>可以标记点，<code>head()</code>就可以取前几个。</p>
<h3
id="降维与聚类主要是找marker基因">降维与聚类，主要是找marker基因</h3>
<h4 id="pca">PCA</h4>
<p>先<code>NormalizeData()</code>，对每个细胞之间的某个基因表达量进行标准化，可以消除不同细胞之间由于测序深度不同而产生的差异，所以归一化后，不同的基因可能仍然有不同的表达量级别。<br />
然后<code>ScaleData()</code>，对每个基因的表达量进行缩放并中心化，可以消除不同基因之间由于表达量级别不同而产生的差异，并且PCA需要基于中心化的数据。</p>
<p>然后就是简单的<code>scRNA &lt;- RunPCA(scRNA, features = VariableFeatures(scRNA))</code>，绘图可以用<code>DimPlot()</code>。</p>
<p>最后挑选主成分，<code>jackStraw()</code>评估每个基因的PCA得分的显著性（对于每个主成分，它会对数据进行PCA，然后对PCA的结果进行随机重排，然后再次进行PCA，然后计算原始PCA（实际）和重排PCA（随机下理论）的差异，这个差异越大，说明这个主成分越重要。），<code>ScoreJackStraw</code>用于分析jackStraw函数的结果，<code>JackStrawPlot()</code>绘制jackStraw图。</p>
这个图，empirical是实际的PCA得分，theoretical是随机重排的PCA得分（这两个对应彩色点），虚线是"均匀分布的
p
值的累积分布函数"，如果主成分轴显着，则二者应显示出差异。<del>不是很清楚，反正选不一样的就完了。</del><br />
这里结果都挺好的，盗用一下某人的例子，像11、12这种就不应该选了： <img
src="https://cdn-ak.f.st-hatena.com/images/fotolife/k/kimoppy126/20180902/20180902233845.png" />
<center>
<a target="_blank" rel="noopener" href="https://www.kimoton.com/entry/20180903/1535903928">《JackStraw
Plot (QQ-plot) を理解する》</a>
</center>
<p>或者可以画<code>ElbowPlot</code>，也可以选择累积解释方差≥90%的主成分数。</p>
<h4 id="聚类还有一堆东西">聚类还有一堆东西</h4>
<p>先<code>FindNeighbors()</code>计算最近邻，然后用<code>FindClusters()</code>聚类。</p>
<p>这里结果可能不太一样（<del>甚至教程的复现和原文不一样</del>），有可能是质控的参数不一样，也有可能作者给出的数据经过质控。</p>
<p><code>FindAllMarkers()</code>找marker基因，真慢啊，有必要换RRO或者python。<br />
完了再筛选出P&lt;0.05的，还有在不同类型细胞中表达量差异大的基因。</p>
<h3 id="qc之二">QC之二</h3>
<p>刚才说的要先聚类或标准化的筛选双细胞和周期判断。</p>
<p><code>emptyDrops()</code>检测空液滴，而检测双细胞可以用<code>doubletFinder()</code>，案例里的<code>doubletCluster()</code>其实寻找的是表达介于两个Cluster之间的Cluster，但他解释的是“有无double
droplet聚在一起的类”（而且好像已经废弃了）。<br />
<code>CellCycleScoring()</code>评估细胞周期。</p>
<h3 id="细胞类型注释-1">细胞类型注释</h3>
<p>按下不表。</p>
<h3 id="轨迹分析-1">轨迹分析</h3>
<p>选择特定基因代表细胞不同阶段的特征，然后用<code>Monocle</code>包，用的2，可以去看他们的文档和论文。</p>
<details>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br></pre></td><td class="code"><pre><span class="line">exprs <span class="operator">&lt;-</span> as<span class="punctuation">(</span>as.matrix<span class="punctuation">(</span>GetAssayData<span class="punctuation">(</span>seu_obj_malOS<span class="punctuation">[[</span><span class="string">&quot;RNA&quot;</span><span class="punctuation">]</span><span class="punctuation">]</span><span class="punctuation">,</span> layer <span class="operator">=</span> <span class="string">&quot;counts&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="string">&#x27;sparseMatrix&#x27;</span><span class="punctuation">)</span></span><br><span class="line">featureData <span class="operator">&lt;-</span> data.frame<span class="punctuation">(</span>gene_short_name <span class="operator">=</span> row.names<span class="punctuation">(</span>seu_obj_malOS<span class="punctuation">)</span><span class="punctuation">,</span>row.names <span class="operator">=</span> row.names<span class="punctuation">(</span>seu_obj_malOS<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">fd <span class="operator">&lt;-</span> new<span class="punctuation">(</span><span class="string">&#x27;AnnotatedDataFrame&#x27;</span><span class="punctuation">,</span> data <span class="operator">=</span> featureData<span class="punctuation">)</span></span><br><span class="line">phenoData <span class="operator">&lt;-</span> seu_obj_malOS<span class="operator">@</span>meta.data</span><br><span class="line">pd <span class="operator">&lt;-</span> new<span class="punctuation">(</span><span class="string">&#x27;AnnotatedDataFrame&#x27;</span><span class="punctuation">,</span> data <span class="operator">=</span> phenoData<span class="punctuation">)</span></span><br><span class="line">cds <span class="operator">&lt;-</span> newCellDataSet<span class="punctuation">(</span>exprs<span class="punctuation">,</span></span><br><span class="line">                      phenoData <span class="operator">=</span> pd<span class="punctuation">,</span> </span><br><span class="line">                      featureData <span class="operator">=</span> fd<span class="punctuation">,</span></span><br><span class="line">                      lowerDetectionLimit <span class="operator">=</span> <span class="number">0.5</span><span class="punctuation">,</span></span><br><span class="line">                      expressionFamily <span class="operator">=</span> negbinomial.size<span class="punctuation">(</span><span class="punctuation">)</span><span class="punctuation">)</span> <span class="comment"># 其实创建cds对象也可以使用importCDS，但是好像seurat V5没法兼容</span></span><br><span class="line"></span><br><span class="line">cds <span class="operator">&lt;-</span> estimateSizeFactors<span class="punctuation">(</span>cds<span class="punctuation">)</span></span><br><span class="line">cds <span class="operator">&lt;-</span> estimateDispersions<span class="punctuation">(</span>cds<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">cds <span class="operator">&lt;-</span> detectGenes<span class="punctuation">(</span>cds<span class="punctuation">,</span> min_expr <span class="operator">=</span> <span class="number">0.1</span><span class="punctuation">)</span></span><br><span class="line">expressed_genes <span class="operator">&lt;-</span> row.names<span class="punctuation">(</span>subset<span class="punctuation">(</span>fData<span class="punctuation">(</span>cds<span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">                                    num_cells_expressed <span class="operator">&gt;=</span> <span class="number">10</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">diff <span class="operator">&lt;-</span> differentialGeneTest<span class="punctuation">(</span>cds<span class="punctuation">[</span>expressed_genes<span class="punctuation">,</span><span class="punctuation">]</span><span class="punctuation">,</span></span><br><span class="line">                             fullModelFormulaStr<span class="operator">=</span><span class="string">&quot;~seurat_clusters&quot;</span><span class="punctuation">,</span></span><br><span class="line">                             cores<span class="operator">=</span><span class="number">10</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">ordergene <span class="operator">&lt;-</span> rownames<span class="punctuation">(</span>subset<span class="punctuation">(</span>diff<span class="punctuation">,</span> qval <span class="operator">&lt;</span> <span class="number">0.01</span><span class="punctuation">)</span><span class="punctuation">)</span> </span><br><span class="line">cds <span class="operator">&lt;-</span> setOrderingFilter<span class="punctuation">(</span>cds<span class="punctuation">,</span> ordergene<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">cds <span class="operator">&lt;-</span> reduceDimension<span class="punctuation">(</span>cds<span class="punctuation">,</span> max_components <span class="operator">=</span> <span class="number">2</span><span class="punctuation">,</span></span><br><span class="line">                       method <span class="operator">=</span> <span class="string">&#x27;DDRTree&#x27;</span><span class="punctuation">)</span></span><br><span class="line">                       </span><br><span class="line">cds <span class="operator">&lt;-</span> orderCells<span class="punctuation">(</span>cds<span class="punctuation">)</span> <span class="comment"># 可能igraph版本不兼容，要换成2.0.3</span></span><br><span class="line"></span><br><span class="line">plot_cell_trajectory<span class="punctuation">(</span>cds<span class="punctuation">,</span>color_by<span class="operator">=</span><span class="string">&quot;Pseudotime&quot;</span><span class="punctuation">,</span> cell_size<span class="operator">=</span><span class="number">1</span><span class="punctuation">,</span>show_backbone<span class="operator">=</span><span class="literal">TRUE</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>
</details>
<h2 id="尾声">尾声</h2>
<h3 id="多样本整合">多样本整合</h3>
<p>用来去掉样本之间的批次效应，可以使用<code>harmony</code>包的<code>RunHarmony()</code>（很强劲），也可以使用<code>IntegrateLayers()</code>。</p>
<h3 id="infecnv"><code>infeCNV</code></h3>
<p>要先准备一份文件。<br />
inferCNVpy快得多，可以试一下。</p>
<details>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br></pre></td><td class="code"><pre><span class="line">celltype <span class="operator">&lt;-</span> cbind<span class="punctuation">(</span>colnames<span class="punctuation">(</span>seu_obj_inferCNV<span class="punctuation">)</span><span class="punctuation">,</span><span class="built_in">as.character</span><span class="punctuation">(</span>seu_obj_inferCNV<span class="operator">@</span>meta.data<span class="operator">$</span>celltype<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">write.table<span class="punctuation">(</span>celltype<span class="punctuation">,</span>file <span class="operator">=</span> <span class="string">&quot;celltype_seu_obj_inferCNV.txt&quot;</span><span class="punctuation">,</span></span><br><span class="line">            sep <span class="operator">=</span> <span class="string">&#x27;\t&#x27;</span><span class="punctuation">,</span>row.names<span class="operator">=</span><span class="built_in">F</span><span class="punctuation">,</span>col.names<span class="operator">=</span><span class="built_in">F</span><span class="punctuation">,</span><span class="built_in">quote</span><span class="operator">=</span><span class="built_in">F</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">mtx <span class="operator">&lt;-</span> as.data.frame<span class="punctuation">(</span>GetAssayData<span class="punctuation">(</span>seu_obj_inferCNV<span class="punctuation">,</span> slot<span class="operator">=</span><span class="string">&#x27;counts&#x27;</span><span class="punctuation">,</span>assay<span class="operator">=</span><span class="string">&#x27;RNA&#x27;</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">geneInfor<span class="operator">=</span>annoGene<span class="punctuation">(</span>rownames<span class="punctuation">(</span>mtx<span class="punctuation">)</span><span class="punctuation">,</span><span class="string">&quot;SYMBOL&quot;</span><span class="punctuation">,</span><span class="string">&#x27;human&#x27;</span><span class="punctuation">)</span></span><br><span class="line">geneInfor_fromEN <span class="operator">=</span> annoGene<span class="punctuation">(</span>rownames<span class="punctuation">(</span>mtx<span class="punctuation">)</span><span class="punctuation">,</span><span class="string">&quot;ENSEMBL&quot;</span><span class="punctuation">,</span><span class="string">&#x27;human&#x27;</span><span class="punctuation">)</span></span><br><span class="line">geneInfor<span class="operator">&lt;-</span>rbind<span class="punctuation">(</span>geneInfor<span class="punctuation">,</span>geneInfor_fromEN<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">ensembl_to_symbol <span class="operator">&lt;-</span> setNames<span class="punctuation">(</span>geneInfor<span class="operator">$</span>SYMBOL<span class="punctuation">,</span> geneInfor<span class="operator">$</span>ENSEMBL<span class="punctuation">)</span></span><br><span class="line">new_rownames <span class="operator">&lt;-</span> rownames<span class="punctuation">(</span>mtx<span class="punctuation">)</span></span><br><span class="line"><span class="keyword">for</span> <span class="punctuation">(</span>i <span class="keyword">in</span> <span class="built_in">seq_along</span><span class="punctuation">(</span>new_rownames<span class="punctuation">)</span><span class="punctuation">)</span> <span class="punctuation">&#123;</span></span><br><span class="line">  <span class="keyword">if</span> <span class="punctuation">(</span>new_rownames<span class="punctuation">[</span>i<span class="punctuation">]</span> <span class="operator">%in%</span> <span class="built_in">names</span><span class="punctuation">(</span>ensembl_to_symbol<span class="punctuation">)</span><span class="punctuation">)</span> <span class="punctuation">&#123;</span></span><br><span class="line">    new_name <span class="operator">&lt;-</span> ensembl_to_symbol<span class="punctuation">[</span>new_rownames<span class="punctuation">[</span>i<span class="punctuation">]</span><span class="punctuation">]</span></span><br><span class="line">    <span class="keyword">if</span> <span class="punctuation">(</span><span class="operator">!</span><span class="punctuation">(</span>new_name <span class="operator">%in%</span> new_rownames<span class="punctuation">)</span><span class="punctuation">)</span> <span class="punctuation">&#123;</span></span><br><span class="line">      new_rownames<span class="punctuation">[</span>i<span class="punctuation">]</span> <span class="operator">&lt;-</span> new_name</span><br><span class="line">    <span class="punctuation">&#125;</span></span><br><span class="line">  <span class="punctuation">&#125;</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br><span class="line">rownames<span class="punctuation">(</span>mtx<span class="punctuation">)</span> <span class="operator">&lt;-</span> new_rownames</span><br><span class="line"></span><br><span class="line">geneInfor<span class="operator">=</span>geneInfor<span class="punctuation">[</span>with<span class="punctuation">(</span>geneInfor<span class="punctuation">,</span> order<span class="punctuation">(</span>chr<span class="punctuation">,</span> start<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span><span class="built_in">c</span><span class="punctuation">(</span><span class="number">1</span><span class="punctuation">,</span><span class="number">4</span><span class="operator">:</span><span class="number">6</span><span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line">geneInfor<span class="operator">=</span>geneInfor<span class="punctuation">[</span><span class="operator">!</span>duplicated<span class="punctuation">(</span>geneInfor<span class="punctuation">[</span><span class="punctuation">,</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">)</span><span class="punctuation">,</span><span class="punctuation">]</span></span><br><span class="line"></span><br><span class="line">geneInfor<span class="operator">$</span>chr_numeric <span class="operator">&lt;-</span> <span class="built_in">as.numeric</span><span class="punctuation">(</span>gsub<span class="punctuation">(</span><span class="string">&quot;chr&quot;</span><span class="punctuation">,</span> <span class="string">&quot;&quot;</span><span class="punctuation">,</span> geneInfor<span class="operator">$</span>chr<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">geneInfor <span class="operator">=</span> geneInfor<span class="punctuation">[</span>order<span class="punctuation">(</span>geneInfor<span class="operator">$</span>chr_numeric<span class="punctuation">,</span> geneInfor<span class="operator">$</span>start<span class="punctuation">)</span><span class="punctuation">,</span> <span class="punctuation">]</span></span><br><span class="line">geneInfor<span class="operator">$</span>chr_numeric <span class="operator">&lt;-</span> <span class="literal">NULL</span></span><br><span class="line">rownames<span class="punctuation">(</span>geneInfor<span class="punctuation">)</span><span class="operator">&lt;-</span><span class="literal">NULL</span></span><br><span class="line"></span><br><span class="line">mtx<span class="operator">=</span>mtx<span class="punctuation">[</span>rownames<span class="punctuation">(</span>mtx<span class="punctuation">)</span> <span class="operator">%in%</span> geneInfor<span class="punctuation">[</span><span class="punctuation">,</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">,</span><span class="punctuation">]</span></span><br><span class="line">mtx<span class="operator">=</span>mtx<span class="punctuation">[</span>match<span class="punctuation">(</span> geneInfor<span class="punctuation">[</span><span class="punctuation">,</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">,</span> rownames<span class="punctuation">(</span>mtx<span class="punctuation">)</span> <span class="punctuation">)</span><span class="punctuation">,</span><span class="punctuation">]</span></span><br><span class="line"></span><br><span class="line">geneFile<span class="operator">=</span><span class="string">&#x27;geneFile.txt&#x27;</span></span><br><span class="line">write.table<span class="punctuation">(</span>geneInfor<span class="punctuation">,</span>file <span class="operator">=</span> geneFile<span class="punctuation">,</span>sep <span class="operator">=</span> <span class="string">&#x27;\t&#x27;</span><span class="punctuation">,</span><span class="built_in">quote</span> <span class="operator">=</span> <span class="built_in">F</span><span class="punctuation">,</span>col.names <span class="operator">=</span> <span class="built_in">F</span><span class="punctuation">,</span>row.names <span class="operator">=</span> <span class="built_in">F</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">infercnv_obj <span class="operator">=</span> CreateInfercnvObject<span class="punctuation">(</span></span><br><span class="line">  raw_counts_matrix <span class="operator">=</span> mtx<span class="punctuation">,</span></span><br><span class="line">  annotations_file <span class="operator">=</span> <span class="string">&#x27;celltype_seu_obj_inferCNV.txt&#x27;</span><span class="punctuation">,</span></span><br><span class="line">  gene_order_file<span class="operator">=</span> geneFile<span class="punctuation">,</span></span><br><span class="line">  ref_group_names<span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;T/NK cells&quot;</span><span class="punctuation">,</span><span class="string">&quot;B cells&quot;</span><span class="punctuation">,</span><span class="string">&quot;Myeloid cells&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span> <span class="comment"># 可以使用免疫细胞作为参考</span></span><br><span class="line"></span><br><span class="line">infercnv_obj <span class="operator">=</span> infercnv<span class="operator">::</span>run<span class="punctuation">(</span>infercnv_obj<span class="punctuation">,</span></span><br><span class="line">                             cutoff<span class="operator">=</span><span class="number">0.1</span><span class="punctuation">,</span> </span><br><span class="line">                             <span class="comment"># use 1 for smart-seq, 0.1 for 10x-genomics</span></span><br><span class="line">                             out_dir <span class="operator">=</span> <span class="string">&#x27;infercnv_large/&#x27;</span><span class="punctuation">,</span><span class="comment"># dir is auto-created for storing outputs</span></span><br><span class="line">                             cluster_by_groups<span class="operator">=</span><span class="built_in">F</span><span class="punctuation">,</span> </span><br><span class="line">                             HMM<span class="operator">=</span><span class="built_in">F</span><span class="punctuation">,</span> plot_steps<span class="operator">=</span><span class="built_in">F</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>
</details>
<h3 id="copykat"><code>copyKAT</code></h3>
<p>这个使用比<code>inferCNV</code>简单得多，就不说了。</p>
<h3 id="细胞通讯">细胞通讯</h3>
<p>可以去看他们的论文。</p>
<details>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br></pre></td><td class="code"><pre><span class="line">data.input <span class="operator">&lt;-</span> seu_obj<span class="punctuation">[[</span><span class="string">&quot;RNA&quot;</span><span class="punctuation">]</span><span class="punctuation">]</span><span class="operator">$</span>data</span><br><span class="line">meta <span class="operator">&lt;-</span> seu_obj<span class="operator">@</span>meta.data<span class="punctuation">[</span><span class="punctuation">,</span><span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;sampleID&quot;</span><span class="punctuation">,</span><span class="string">&quot;celltype&quot;</span><span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line">colnames<span class="punctuation">(</span>meta<span class="punctuation">)</span> <span class="operator">&lt;-</span>  <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;samples&quot;</span><span class="punctuation">,</span><span class="string">&quot;celltypes&quot;</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">meta<span class="operator">$</span>celltypes <span class="operator">&lt;-</span> gsub<span class="punctuation">(</span><span class="string">&quot; cells&quot;</span><span class="punctuation">,</span> <span class="string">&quot;&quot;</span><span class="punctuation">,</span> meta<span class="operator">$</span>celltypes<span class="punctuation">)</span></span><br><span class="line">table<span class="punctuation">(</span>meta<span class="operator">$</span>celltypes<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">ordered_indices <span class="operator">&lt;-</span> order<span class="punctuation">(</span>meta<span class="operator">$</span>celltypes<span class="punctuation">)</span></span><br><span class="line">meta <span class="operator">&lt;-</span> meta<span class="punctuation">[</span>ordered_indices<span class="punctuation">,</span> <span class="punctuation">]</span></span><br><span class="line">data.input <span class="operator">&lt;-</span> data.input<span class="punctuation">[</span><span class="punctuation">,</span> ordered_indices<span class="punctuation">]</span></span><br><span class="line">identical<span class="punctuation">(</span>rownames<span class="punctuation">(</span>meta<span class="punctuation">)</span><span class="punctuation">,</span>colnames<span class="punctuation">(</span>data.input<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">cellchat <span class="operator">&lt;-</span> createCellChat<span class="punctuation">(</span>object <span class="operator">=</span> data.input<span class="punctuation">,</span> </span><br><span class="line">                           meta <span class="operator">=</span> meta<span class="punctuation">,</span> </span><br><span class="line">                           group.by <span class="operator">=</span> <span class="string">&quot;celltypes&quot;</span><span class="punctuation">)</span></span><br><span class="line">CellChatDB <span class="operator">&lt;-</span> CellChatDB.human</span><br><span class="line">showDatabaseCategory<span class="punctuation">(</span>CellChatDB<span class="punctuation">)</span></span><br><span class="line">dplyr<span class="operator">::</span>glimpse<span class="punctuation">(</span>CellChatDB<span class="operator">$</span>interaction<span class="punctuation">)</span></span><br><span class="line">CellChatDB.use <span class="operator">&lt;-</span> CellChatDB</span><br><span class="line"></span><br><span class="line">cellchat<span class="operator">@</span>DB <span class="operator">&lt;-</span> CellChatDB.use</span><br><span class="line">cellchat <span class="operator">&lt;-</span> subsetData<span class="punctuation">(</span>cellchat<span class="punctuation">)</span> </span><br><span class="line"></span><br><span class="line">options<span class="punctuation">(</span>future.globals.maxSize<span class="operator">=</span> <span class="number">10000</span><span class="operator">*</span><span class="number">1024</span><span class="operator">^</span><span class="number">2</span><span class="punctuation">)</span></span><br><span class="line">future<span class="operator">::</span>plan<span class="punctuation">(</span><span class="string">&quot;multisession&quot;</span><span class="punctuation">,</span> workers <span class="operator">=</span> <span class="number">10</span><span class="punctuation">)</span></span><br><span class="line">cellchat <span class="operator">&lt;-</span> identifyOverExpressedGenes<span class="punctuation">(</span>cellchat<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">cellchat <span class="operator">&lt;-</span> identifyOverExpressedInteractions<span class="punctuation">(</span>cellchat<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">cellchat <span class="operator">&lt;-</span> computeCommunProb<span class="punctuation">(</span>cellchat<span class="punctuation">,</span> type <span class="operator">=</span> <span class="string">&quot;triMean&quot;</span><span class="punctuation">,</span>raw.use <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">cellchat <span class="operator">&lt;-</span> filterCommunication<span class="punctuation">(</span>cellchat<span class="punctuation">,</span> min.cells <span class="operator">=</span> <span class="number">10</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">cellchat <span class="operator">&lt;-</span> computeCommunProbPathway<span class="punctuation">(</span>cellchat<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">cellchat <span class="operator">&lt;-</span> aggregateNet<span class="punctuation">(</span>cellchat<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">groupSize <span class="operator">&lt;-</span> <span class="built_in">as.numeric</span><span class="punctuation">(</span>table<span class="punctuation">(</span>cellchat<span class="operator">@</span>idents<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">netVisual_circle<span class="punctuation">(</span>cellchat<span class="operator">@</span>net<span class="operator">$</span>count<span class="punctuation">,</span> vertex.weight <span class="operator">=</span> groupSize<span class="punctuation">,</span></span><br><span class="line">                 weight.scale <span class="operator">=</span> <span class="built_in">T</span><span class="punctuation">,</span> label.edge<span class="operator">=</span> <span class="built_in">F</span><span class="punctuation">,</span> title.name <span class="operator">=</span> <span class="string">&quot;Number of interactions&quot;</span><span class="punctuation">)</span></span><br><span class="line">netVisual_circle<span class="punctuation">(</span>cellchat<span class="operator">@</span>net<span class="operator">$</span>weight<span class="punctuation">,</span> vertex.weight <span class="operator">=</span> groupSize<span class="punctuation">,</span> </span><br><span class="line">                 weight.scale <span class="operator">=</span> <span class="built_in">T</span><span class="punctuation">,</span> label.edge<span class="operator">=</span> <span class="built_in">F</span><span class="punctuation">,</span> title.name <span class="operator">=</span> <span class="string">&quot;Interaction weights/strength&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>
</details>
<p>有时间再用Python实现一下。</p>
<h2 id="python实现">Python实现</h2>
<p>其实就这些流程化的步骤，原理在<code>scanpy</code>中是一样的，见<a
target="_blank" rel="noopener" href="https://scanpy.readthedocs.io/en/stable/tutorials/basics/clustering-2017.html">他们的教程</a>。</p>
<p>但是有一点好玩的，能用聚类结果绘制paga图，然后作为<code>scanpy.tl.umap()</code>的初始化参数，可以解决其它特征空间聚类得到的簇在UMAP空间中明显不相连的问题。</p>
<p><font color="red"><strong>おわり</strong></font></p>

        
      </div>

         
    </div>
    
     
  </div>
  
    
<nav id="article-nav">
  <a class="article-nav-btn left "
    
      href="/blog/2024/08/01/reproduce-10-1038-slash-s41586-024-07504-y/"
      title="复现：Single-cell and spatial atlases of spinal cord injury in the Tabulae Paralytica"
     >
    <i class="fa-solid fa-angle-left"></i>
    <p class="title-text">
      
        复现：Single-cell and spatial atlases of spinal cord injury in the Tabulae Paralytica
        
    </p>
  </a>
  <a class="article-nav-btn right "
    
      href="/blog/2024/03/28/page-clock/"
      title="一种实现翻页式时钟的方法"
     >

    <p class="title-text">
      
        一种实现翻页式时钟的方法
        
    </p>
    <i class="fa-solid fa-angle-right"></i>
  </a>
</nav>


  
</article>




    </div>
    <div id="footer-wrapper">
      <footer id="footer">
  
  <div id="footer-info" class="inner">
    
    &copy; 2025 kusurin<br>
    Powered by <a href="https://hexo.io/" target="_blank">Hexo</a> & Theme <a target="_blank" rel="noopener" href="https://github.com/saicaca/hexo-theme-vivia">Vivia</a>
  </div>
</footer>

    </div>
    <div class="back-to-top-wrapper">
    <button id="back-to-top-btn" class="back-to-top-btn hide" onclick="topFunction()">
        <i class="fa-solid fa-angle-up"></i>
    </button>
</div>

<script>
    function topFunction() {
        window.scroll({ top: 0, behavior: 'smooth' });
    }
    let btn = document.getElementById('back-to-top-btn');
    function scrollFunction() {
        if (document.body.scrollTop > 600 || document.documentElement.scrollTop > 600) {
            btn.classList.remove('hide')
        } else {
            btn.classList.add('hide')
        }
    }
    window.onscroll = function() {
        scrollFunction();
    }
</script>

  </div>
  <script src="/blog/js/light-dark-switch.js"></script>

</body>
</html>
